{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d1a2ead4",
   "metadata": {},
   "source": [
    "# Project Goals:\n",
    "- Create documented files to clean and prepare H4 Nucleosome Occupancy dataset for processing by modeling via a multiple classification ML and ANN algorithms.\n",
    "- Use K-mers to create datasets from the raw genetic sequence data\n",
    "- Pass K-mer datasets to a Bag of Words CountVectorizer\n",
    "- Pass dataframes with new vectorized features and nucleotide counts to classification models.\n",
    "- Find a model that hopefully outpreforms both the modal baseline and the best preforming model from the origin website\n",
    "\n",
    "# Executive Summary:\n",
    "> ## Key Takeaways:\n",
    "> - I achieved a ~84.4% accuracy on test data using a Multinomial Naive Bayes test.\n",
    "> - I did beat baseline but did not beat the best model from www.jaist.ac\n",
    ">     - My best model lost by roughly 4 points\n",
    "\n",
    "> ## Reccomendations:\n",
    "> I reccomend that the model used by www.jaist.ac be used in lieu of my own models. Additionally, the hexamer dataset seemed to be more indicative of positive or negative labels than the codon dataset. If pursuing the ANN models further I reccomend using a model with at least one hidden layer as it outpreformed the single layer model slightly with the codon dataset and significantly improved with the hexamer dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0585ef7e",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9e46ee2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import acquire as ac\n",
    "import preprocessing as pr \n",
    "import explore as ex\n",
    "import model as md\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import unicodedata\n",
    "import nltk\n",
    "\n",
    "from tensorflow import keras\n",
    "import tensorflow as tf\n",
    "from keras.regularizers import l2\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "import os\n",
    "from wordcloud import WordCloud\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import math\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "#Importing different classifier from sklearn\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn import svm\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "from sklearn.naive_bayes import MultinomialNB\n",
    "from sklearn.gaussian_process.kernels import RBF\n",
    "from sklearn.gaussian_process import GaussianProcessClassifier\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier\n",
    "from sklearn.metrics import classification_report, accuracy_score\n",
    "from sklearn.model_selection import KFold, cross_val_score, cross_validate\n",
    "\n",
    "import warnings\n",
    "warnings.simplefilter('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85ac0172",
   "metadata": {},
   "source": [
    "# Acquire\n",
    "#### With the help of my acquire.py script:\n",
    "- Pull H4 nucleosome occupancy sequence data from http://www.jaist.ac.jp/~tran/nucleosome/ten_dataset/dataset02.txt in the form of a .txt file\n",
    "- Convert this .txt file into a format useable by pandas for passing to dataframe\n",
    "- Preform pre-preperation summary of data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3f48d0c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>iTELL-Chr1_61\\n</td>\n",
       "      <td>CACACCACACCCACACACCCACACACCACACCACACACCACACCAC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>iYAL067W-A_3174\\n</td>\n",
       "      <td>TTTCTAAGTGGGATTTTTCTTAATCCTTGGATTCTTAAAAGGTTAT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>iYAL067W-A_4141\\n</td>\n",
       "      <td>AAAGCCATTAAGGTCCAGATAACCAAGGGACAATAAACCTATGCTT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id                                           sequence  label\n",
       "0    iTELL-Chr1_61\\n  CACACCACACCCACACACCCACACACCACACCACACACCACACCAC...      0\n",
       "1  iYAL067W-A_3174\\n  TTTCTAAGTGGGATTTTTCTTAATCCTTGGATTCTTAAAAGGTTAT...      1\n",
       "2  iYAL067W-A_4141\\n  AAAGCCATTAAGGTCCAGATAACCAAGGGACAATAAACCTATGCTT...      1"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Acquire raw data on histone occupancy\n",
    "h4 = ac.get_H4_data()\n",
    "h4.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bb61df92",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(id          0\n",
       " sequence    0\n",
       " label       0\n",
       " dtype: int64,\n",
       " (14601, 3))"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking for nulls and shape of data\n",
    "h4.isnull().sum(), h4.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9f6468d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjEAAAGsCAYAAAAsf/b0AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA1DUlEQVR4nO3dfXRU9Z3H8c+YhyFJkytJyAyzREhpSsGAtcENiVrQEB5KTFu0aNPNwRV5WCw0Shak7Gr0tInQCrRNpUgpIA/ibo9pbbEpobbUHBJAbKxBpLbyWBiCNkwCpAmEu394uOsQniYg4Ze8X+fcc5zf/d57f79MrvPhN/feuGzbtgUAAGCYGzq7AwAAAB1BiAEAAEYixAAAACMRYgAAgJEIMQAAwEiEGAAAYCRCDAAAMBIhBgAAGIkQAwAAjESIAT5BK1eulMvlcpYePXrI6/XqrrvuUmlpqerr69ttU1xcLJfLFdJxTp48qeLiYv3hD38IabvzHatfv37Kzc0NaT+Xsm7dOi1evPi861wul4qLi6/q8a623/3udxo6dKhiYmLkcrn0i1/8ol3NiBEjgt7rCy3X21g7+rsDXA/CO7sDQHewYsUKfe5zn9OpU6dUX1+vqqoqzZ8/X9///vf10ksvaeTIkU7tww8/rDFjxoS0/5MnT+qpp56S9NGH6eXqyLE6Yt26daqrq1NhYWG7ddXV1erTp88n3oeOsm1bEyZM0Gc/+1m98soriomJ0YABA9rVPffcc2psbHReb9iwQd/5znec9/6s622sHf3dAa4HhBjgGkhLS9PQoUOd1/fee68effRR3XHHHRo/frzee+89eTweSR99yH3SH3QnT55UdHT0NTnWpQwbNqxTj38phw4d0j/+8Q999atfVXZ29gXrBg0aFPT63XffldT+ve+os+8ZgP/H10lAJ7npppv07LPPqqmpSUuXLnXaz/cVz2uvvaYRI0YoISFBUVFRuummm3Tvvffq5MmT2rt3r3r16iVJeuqpp5yvLR588MGg/b355pu677771LNnT/Xv3/+CxzqrvLxcQ4YMUY8ePfTpT39aP/zhD4PWn/2qbO/evUHtf/jDH+RyuZyvJ0aMGKENGzZo3759QV+rnHW+r1jq6ur05S9/WT179lSPHj30+c9/XqtWrTrvcV588UXNmzdPPp9PcXFxGjlypHbv3n3hH/zHVFVVKTs7W7GxsYqOjlZWVpY2bNjgrC8uLnZC3pw5c+RyudSvX7/L2vf5VFZW6stf/rL69OmjHj166DOf+YymTp2qDz74IKjuYu9ZS0uLZs2aJa/Xq+joaH3xi1/Ujh071K9fP+c9P8vv92vq1Knq06ePIiMjlZKSoqeeekqnT5+WpEv+7gDXO2ZigE70pS99SWFhYfrjH/94wZq9e/dq3LhxuvPOO/Wzn/1MN954o/7+97+roqJCra2t6t27tyoqKjRmzBhNmjRJDz/8sCQ5H05njR8/Xg888ICmTZumEydOXLRftbW1KiwsVHFxsbxer9auXatvfetbam1tVVFRUUhjfO655zRlyhT97W9/U3l5+SXrd+/eraysLCUlJemHP/yhEhIStGbNGj344IM6cuSIZs+eHVT/7W9/W7fffrt++tOfqrGxUXPmzNE999yjXbt2KSws7ILH2bx5s3JycjRkyBAtX75cbrdbzz33nO655x69+OKLuv/++/Xwww/rlltu0fjx4zVjxgzl5+fL7XaHNP6P+9vf/qbMzEw9/PDDsixLe/fu1cKFC3XHHXfo7bffVkRERFD9+d6zf//3f9dLL72k2bNn6+6779Y777yjr371q0FfZUkfBZh//dd/1Q033KAnnnhC/fv3V3V1tb7zne9o7969WrFixWX/7gDXLRvAJ2bFihW2JHv79u0XrPF4PPbAgQOd108++aT98VPz5z//uS3Jrq2tveA+jh49akuyn3zyyXbrzu7viSeeuOC6j+vbt6/tcrnaHS8nJ8eOi4uzT5w4ETS2PXv2BNX9/ve/tyXZv//97522cePG2X379j1v38/t9wMPPGC73W57//79QXVjx461o6Oj7WPHjgUd50tf+lJQ3f/8z//Ykuzq6urzHu+sYcOG2UlJSXZTU5PTdvr0aTstLc3u06ePfebMGdu2bXvPnj22JPt73/veRfd3rku992fOnLFPnTpl79u3z5Zk//KXv3TWXeg927lzpy3JnjNnTlD7iy++aEuyJ06c6LRNnTrV/tSnPmXv27cvqPb73/++LcneuXOnbdsX/90Brnd8nQR0Mtu2L7r+85//vCIjIzVlyhStWrVK77//foeOc++991527c0336xbbrklqC0/P1+NjY168803O3T8y/Xaa68pOztbycnJQe0PPvigTp48qerq6qD2vLy8oNdDhgyRJO3bt++Cxzhx4oS2bt2q++67T5/61Kec9rCwMBUUFOjgwYOX/ZVUKOrr6zVt2jQlJycrPDxcERER6tu3ryRp165d7erPfc82b94sSZowYUJQ+3333afw8OCJ9V//+te666675PP5dPr0aWcZO3Zs0L4AkxFigE504sQJffjhh/L5fBes6d+/vzZt2qSkpCQ98sgj6t+/v/r3768f/OAHIR2rd+/el13r9Xov2Pbhhx+GdNxQffjhh+ft69mf0bnHT0hICHp99uue5ubmCx6joaFBtm2HdJwrdebMGY0aNUovv/yyZs+erd/97nfatm2bampqLtjfc/t3tk9nLwI/Kzw8vN3P4ciRI/rVr36liIiIoOXmm2+WpHbX4QAm4poYoBNt2LBBbW1tl7y19c4779Sdd96ptrY2vfHGG/rRj36kwsJCeTwePfDAA5d1rFCePeP3+y/YdvbDskePHpI+utD04670wzEhIUGHDx9u137o0CFJUmJi4hXtX5J69uypG2644RM/zsfV1dXprbfe0sqVKzVx4kSn/a9//esFtzn3PTv7sz9y5Ij+5V/+xWk/ffp0u9CVmJioIUOG6Lvf/e55932x4AyYgpkYoJPs379fRUVFsixLU6dOvaxtwsLClJGRoR//+MeS5Hy1czmzD6HYuXOn3nrrraC2devWKTY2Vl/4whckyblL589//nNQ3SuvvNJuf263+7L7lp2drddee80JE2e98MILio6Oviq3ZMfExCgjI0Mvv/xyUL/OnDmjNWvWqE+fPvrsZz97xcf5uLOB5NwLgz9+Z9qlfPGLX5QkvfTSS0HtP//5z507js7Kzc1VXV2d+vfvr6FDh7ZbzoaYq/27A1xLzMQA10BdXZ1zTUJ9fb1ef/11rVixQmFhYSovL7/o3SA/+clP9Nprr2ncuHG66aab9M9//lM/+9nPJMl5SF5sbKz69u2rX/7yl8rOzlZ8fLwSExM7fDuwz+dTXl6eiouL1bt3b61Zs0aVlZWaP3++86yS2267TQMGDFBRUZFOnz6tnj17qry8XFVVVe32N3jwYL388stasmSJ0tPTdcMNN1zw2SlPPvmkcz3HE088ofj4eK1du1YbNmzQggULZFlWh8Z0rtLSUuXk5Oiuu+5SUVGRIiMj9dxzz6murk4vvvhiyE9NvpTPfe5z6t+/vx5//HHZtq34+Hj96le/UmVl5WXv4+abb9bXv/51PfvsswoLC9Pdd9+tnTt36tlnn5VlWbrhhv//d+nTTz+tyspKZWVlaebMmRowYID++c9/au/evXr11Vf1k5/8RH369LnqvzvANdXJFxYDXdrZO1TOLpGRkXZSUpI9fPhwu6SkxK6vr2+3zbl3DFVXV9tf/epX7b59+9put9tOSEiwhw8fbr/yyitB223atMm+9dZbbbfbHXSnytn9HT169JLHsu2P7k4aN26c/fOf/9y++eab7cjISLtfv372woUL223/l7/8xR41apQdFxdn9+rVy54xY4a9YcOGdncn/eMf/7Dvu+8++8Ybb7RdLlfQMXWeO2Pefvtt+5577rEty7IjIyPtW265xV6xYkVQzdm7k/73f/83qP3s3UTn1p/P66+/bt999912TEyMHRUVZQ8bNsz+1a9+dd79XY27k9555x07JyfHjo2NtXv27Gl/7Wtfs/fv39/uZ3Cx9+yf//yn/dhjj9lJSUl2jx497GHDhtnV1dW2ZVn2o48+GlR79OhRe+bMmXZKSoodERFhx8fH2+np6fa8efPs48ePO3UX+t0Brncu277ErREAgOvali1bdPvtt2vt2rXKz8/v7O4A1wwhBgAMUllZqerqaqWnpysqKkpvvfWWnnnmGVmWpT//+c/OBddAd8A1MQBgkLi4OG3cuFGLFy9WU1OTEhMTNXbsWJWWlhJg0O0wEwMAAIzELdYAAMBIhBgAAGAkQgwAADBSl72w98yZMzp06JBiY2Ov+kOrAADAJ8O2bTU1Ncnn8wU9wPF8umyIOXToULu/ggsAAMxw4MAB9enT56I1XTbExMbGSvrohxAXF9fJvQEAAJejsbFRycnJzuf4xXTZEHP2K6S4uDhCDAAAhrmcS0G4sBcAABiJEAMAAIxEiAEAAEYixAAAACMRYgAAgJEIMQAAwEiEGAAAYCRCDAAAMBIhBgAAGIkQAwAAjESIAQAARiLEAAAAIxFiAACAkQgxAADASOGd3QFcff0e39DZXcA1tPeZcZ3dBQDoFMzEAAAAI4UUYk6fPq3/+q//UkpKiqKiovTpT39aTz/9tM6cOePU2Lat4uJi+Xw+RUVFacSIEdq5c2fQflpaWjRjxgwlJiYqJiZGeXl5OnjwYFBNQ0ODCgoKZFmWLMtSQUGBjh071vGRAgCALiWkEDN//nz95Cc/UVlZmXbt2qUFCxboe9/7nn70ox85NQsWLNDChQtVVlam7du3y+v1KicnR01NTU5NYWGhysvLtX79elVVVen48ePKzc1VW1ubU5Ofn6/a2lpVVFSooqJCtbW1KigouApDBgAAXYHLtm37cotzc3Pl8Xi0fPlyp+3ee+9VdHS0Vq9eLdu25fP5VFhYqDlz5kj6aNbF4/Fo/vz5mjp1qgKBgHr16qXVq1fr/vvvlyQdOnRIycnJevXVVzV69Gjt2rVLgwYNUk1NjTIyMiRJNTU1yszM1LvvvqsBAwZcsq+NjY2yLEuBQEBxcXEh/VBMxzUx3QvXxADoSkL5/A5pJuaOO+7Q7373O/3lL3+RJL311luqqqrSl770JUnSnj175Pf7NWrUKGcbt9ut4cOHa8uWLZKkHTt26NSpU0E1Pp9PaWlpTk11dbUsy3ICjCQNGzZMlmU5NedqaWlRY2Nj0AIAALqukO5OmjNnjgKBgD73uc8pLCxMbW1t+u53v6uvf/3rkiS/3y9J8ng8Qdt5PB7t27fPqYmMjFTPnj3b1Zzd3u/3Kykpqd3xk5KSnJpzlZaW6qmnngplOAAAwGAhzcS89NJLWrNmjdatW6c333xTq1at0ve//32tWrUqqM7lcgW9tm27Xdu5zq05X/3F9jN37lwFAgFnOXDgwOUOCwAAGCikmZj//M//1OOPP64HHnhAkjR48GDt27dPpaWlmjhxorxer6SPZlJ69+7tbFdfX+/Mzni9XrW2tqqhoSFoNqa+vl5ZWVlOzZEjR9od/+jRo+1mec5yu91yu92hDAcAABgspJmYkydP6oYbgjcJCwtzbrFOSUmR1+tVZWWls761tVWbN292Akp6eroiIiKCag4fPqy6ujqnJjMzU4FAQNu2bXNqtm7dqkAg4NQAAIDuLaSZmHvuuUff/e53ddNNN+nmm2/Wn/70Jy1cuFAPPfSQpI++AiosLFRJSYlSU1OVmpqqkpISRUdHKz8/X5JkWZYmTZqkWbNmKSEhQfHx8SoqKtLgwYM1cuRISdLAgQM1ZswYTZ48WUuXLpUkTZkyRbm5uZd1ZxIAAOj6QgoxP/rRj/Tf//3fmj59uurr6+Xz+TR16lQ98cQTTs3s2bPV3Nys6dOnq6GhQRkZGdq4caNiY2OdmkWLFik8PFwTJkxQc3OzsrOztXLlSoWFhTk1a9eu1cyZM527mPLy8lRWVnal4wUAAF1ESM+JMQnPiUF3wXNiAHQln9hzYgAAAK4XhBgAAGAkQgwAADASIQYAABiJEAMAAIxEiAEAAEYixAAAACMRYgAAgJEIMQAAwEiEGAAAYCRCDAAAMBIhBgAAGIkQAwAAjESIAQAARiLEAAAAIxFiAACAkQgxAADASIQYAABgJEIMAAAwEiEGAAAYiRADAACMRIgBAABGIsQAAAAjEWIAAICRCDEAAMBIhBgAAGAkQgwAADASIQYAABiJEAMAAIxEiAEAAEYixAAAACMRYgAAgJFCCjH9+vWTy+VqtzzyyCOSJNu2VVxcLJ/Pp6ioKI0YMUI7d+4M2kdLS4tmzJihxMRExcTEKC8vTwcPHgyqaWhoUEFBgSzLkmVZKigo0LFjx65spAAAoEsJKcRs375dhw8fdpbKykpJ0te+9jVJ0oIFC7Rw4UKVlZVp+/bt8nq9ysnJUVNTk7OPwsJClZeXa/369aqqqtLx48eVm5urtrY2pyY/P1+1tbWqqKhQRUWFamtrVVBQcDXGCwAAugiXbdt2RzcuLCzUr3/9a7333nuSJJ/Pp8LCQs2ZM0fSR7MuHo9H8+fP19SpUxUIBNSrVy+tXr1a999/vyTp0KFDSk5O1quvvqrRo0dr165dGjRokGpqapSRkSFJqqmpUWZmpt59910NGDDgsvrW2Ngoy7IUCAQUFxfX0SEaqd/jGzq7C7iG9j4zrrO7AABXTSif3x2+Jqa1tVVr1qzRQw89JJfLpT179sjv92vUqFFOjdvt1vDhw7VlyxZJ0o4dO3Tq1KmgGp/Pp7S0NKemurpalmU5AUaShg0bJsuynJrzaWlpUWNjY9ACAAC6rg6HmF/84hc6duyYHnzwQUmS3++XJHk8nqA6j8fjrPP7/YqMjFTPnj0vWpOUlNTueElJSU7N+ZSWljrX0FiWpeTk5I4ODQAAGKDDIWb58uUaO3asfD5fULvL5Qp6bdt2u7ZznVtzvvpL7Wfu3LkKBALOcuDAgcsZBgAAMFSHQsy+ffu0adMmPfzww06b1+uVpHazJfX19c7sjNfrVWtrqxoaGi5ac+TIkXbHPHr0aLtZno9zu92Ki4sLWgAAQNfVoRCzYsUKJSUlady4/7+gMCUlRV6v17ljSfroupnNmzcrKytLkpSenq6IiIigmsOHD6uurs6pyczMVCAQ0LZt25yarVu3KhAIODUAAADhoW5w5swZrVixQhMnTlR4+P9v7nK5VFhYqJKSEqWmpio1NVUlJSWKjo5Wfn6+JMmyLE2aNEmzZs1SQkKC4uPjVVRUpMGDB2vkyJGSpIEDB2rMmDGaPHmyli5dKkmaMmWKcnNzL/vOJAAA0PWFHGI2bdqk/fv366GHHmq3bvbs2Wpubtb06dPV0NCgjIwMbdy4UbGxsU7NokWLFB4ergkTJqi5uVnZ2dlauXKlwsLCnJq1a9dq5syZzl1MeXl5Kisr68j4AABAF3VFz4m5nvGcGHQXPCcGQFdyTZ4TAwAA0JkIMQAAwEiEGAAAYCRCDAAAMBIhBgAAGIkQAwAAjESIAQAARiLEAAAAIxFiAACAkQgxAADASIQYAABgJEIMAAAwEiEGAAAYiRADAACMRIgBAABGIsQAAAAjhXd2BwAAl6/f4xs6uwu4hvY+M66zu3BdYyYGAAAYiRADAACMRIgBAABGIsQAAAAjEWIAAICRCDEAAMBIhBgAAGAkQgwAADASIQYAABiJEAMAAIxEiAEAAEYixAAAACMRYgAAgJEIMQAAwEiEGAAAYKSQQ8zf//53/du//ZsSEhIUHR2tz3/+89qxY4ez3rZtFRcXy+fzKSoqSiNGjNDOnTuD9tHS0qIZM2YoMTFRMTExysvL08GDB4NqGhoaVFBQIMuyZFmWCgoKdOzYsY6NEgAAdDkhhZiGhgbdfvvtioiI0G9+8xu98847evbZZ3XjjTc6NQsWLNDChQtVVlam7du3y+v1KicnR01NTU5NYWGhysvLtX79elVVVen48ePKzc1VW1ubU5Ofn6/a2lpVVFSooqJCtbW1KigouPIRAwCALiE8lOL58+crOTlZK1ascNr69evn/Ldt21q8eLHmzZun8ePHS5JWrVolj8ejdevWaerUqQoEAlq+fLlWr16tkSNHSpLWrFmj5ORkbdq0SaNHj9auXbtUUVGhmpoaZWRkSJKWLVumzMxM7d69WwMGDLjScQMAAMOFNBPzyiuvaOjQofra176mpKQk3XrrrVq2bJmzfs+ePfL7/Ro1apTT5na7NXz4cG3ZskWStGPHDp06dSqoxufzKS0tzamprq6WZVlOgJGkYcOGybIsp+ZcLS0tamxsDFoAAEDXFVKIef/997VkyRKlpqbqt7/9raZNm6aZM2fqhRdekCT5/X5JksfjCdrO4/E46/x+vyIjI9WzZ8+L1iQlJbU7flJSklNzrtLSUuf6GcuylJycHMrQAACAYUIKMWfOnNEXvvAFlZSU6NZbb9XUqVM1efJkLVmyJKjO5XIFvbZtu13buc6tOV/9xfYzd+5cBQIBZzlw4MDlDgsAABgopBDTu3dvDRo0KKht4MCB2r9/vyTJ6/VKUrvZkvr6emd2xuv1qrW1VQ0NDRetOXLkSLvjHz16tN0sz1lut1txcXFBCwAA6LpCCjG33367du/eHdT2l7/8RX379pUkpaSkyOv1qrKy0lnf2tqqzZs3KysrS5KUnp6uiIiIoJrDhw+rrq7OqcnMzFQgENC2bducmq1btyoQCDg1AACgewvp7qRHH31UWVlZKikp0YQJE7Rt2zY9//zzev755yV99BVQYWGhSkpKlJqaqtTUVJWUlCg6Olr5+fmSJMuyNGnSJM2aNUsJCQmKj49XUVGRBg8e7NytNHDgQI0ZM0aTJ0/W0qVLJUlTpkxRbm4udyYBAABJIYaY2267TeXl5Zo7d66efvpppaSkaPHixfrGN77h1MyePVvNzc2aPn26GhoalJGRoY0bNyo2NtapWbRokcLDwzVhwgQ1NzcrOztbK1euVFhYmFOzdu1azZw507mLKS8vT2VlZVc6XgAA0EW4bNu2O7sTn4TGxkZZlqVAINDtro/p9/iGzu4CrqG9z4zr7C7gGuL87l664/kdyuc3fzsJAAAYiRADAACMRIgBAABGIsQAAAAjEWIAAICRCDEAAMBIhBgAAGAkQgwAADASIQYAABiJEAMAAIxEiAEAAEYixAAAACMRYgAAgJEIMQAAwEiEGAAAYCRCDAAAMBIhBgAAGIkQAwAAjESIAQAARiLEAAAAIxFiAACAkQgxAADASIQYAABgJEIMAAAwEiEGAAAYiRADAACMRIgBAABGIsQAAAAjEWIAAICRCDEAAMBIhBgAAGAkQgwAADASIQYAABgppBBTXFwsl8sVtHi9Xme9bdsqLi6Wz+dTVFSURowYoZ07dwbto6WlRTNmzFBiYqJiYmKUl5engwcPBtU0NDSooKBAlmXJsiwVFBTo2LFjHR8lAADockKeibn55pt1+PBhZ3n77beddQsWLNDChQtVVlam7du3y+v1KicnR01NTU5NYWGhysvLtX79elVVVen48ePKzc1VW1ubU5Ofn6/a2lpVVFSooqJCtbW1KigouMKhAgCAriQ85A3Cw4NmX86ybVuLFy/WvHnzNH78eEnSqlWr5PF4tG7dOk2dOlWBQEDLly/X6tWrNXLkSEnSmjVrlJycrE2bNmn06NHatWuXKioqVFNTo4yMDEnSsmXLlJmZqd27d2vAgAFXMl4AANBFhDwT895778nn8yklJUUPPPCA3n//fUnSnj175Pf7NWrUKKfW7XZr+PDh2rJliyRpx44dOnXqVFCNz+dTWlqaU1NdXS3LspwAI0nDhg2TZVlOzfm0tLSosbExaAEAAF1XSCEmIyNDL7zwgn77299q2bJl8vv9ysrK0ocffii/3y9J8ng8Qdt4PB5nnd/vV2RkpHr27HnRmqSkpHbHTkpKcmrOp7S01LmGxrIsJScnhzI0AABgmJBCzNixY3Xvvfdq8ODBGjlypDZs2CDpo6+NznK5XEHb2Lbdru1c59acr/5S+5k7d64CgYCzHDhw4LLGBAAAzHRFt1jHxMRo8ODBeu+995zrZM6dLamvr3dmZ7xer1pbW9XQ0HDRmiNHjrQ71tGjR9vN8nyc2+1WXFxc0AIAALquKwoxLS0t2rVrl3r37q2UlBR5vV5VVlY661tbW7V582ZlZWVJktLT0xURERFUc/jwYdXV1Tk1mZmZCgQC2rZtm1OzdetWBQIBpwYAACCku5OKiop0zz336KabblJ9fb2+853vqLGxURMnTpTL5VJhYaFKSkqUmpqq1NRUlZSUKDo6Wvn5+ZIky7I0adIkzZo1SwkJCYqPj1dRUZHz9ZQkDRw4UGPGjNHkyZO1dOlSSdKUKVOUm5vLnUkAAMARUog5ePCgvv71r+uDDz5Qr169NGzYMNXU1Khv376SpNmzZ6u5uVnTp09XQ0ODMjIytHHjRsXGxjr7WLRokcLDwzVhwgQ1NzcrOztbK1euVFhYmFOzdu1azZw507mLKS8vT2VlZVdjvAAAoItw2bZtd3YnPgmNjY2yLEuBQKDbXR/T7/ENnd0FXEN7nxnX2V3ANcT53b10x/M7lM9v/nYSAAAwEiEGAAAYiRADAACMRIgBAABGIsQAAAAjEWIAAICRCDEAAMBIhBgAAGAkQgwAADASIQYAABiJEAMAAIxEiAEAAEYixAAAACMRYgAAgJEIMQAAwEiEGAAAYCRCDAAAMBIhBgAAGIkQAwAAjESIAQAARiLEAAAAIxFiAACAkQgxAADASIQYAABgJEIMAAAwEiEGAAAYiRADAACMRIgBAABGIsQAAAAjEWIAAICRCDEAAMBIhBgAAGCkKwoxpaWlcrlcKiwsdNps21ZxcbF8Pp+ioqI0YsQI7dy5M2i7lpYWzZgxQ4mJiYqJiVFeXp4OHjwYVNPQ0KCCggJZliXLslRQUKBjx45dSXcBAEAX0uEQs337dj3//PMaMmRIUPuCBQu0cOFClZWVafv27fJ6vcrJyVFTU5NTU1hYqPLycq1fv15VVVU6fvy4cnNz1dbW5tTk5+ertrZWFRUVqqioUG1trQoKCjraXQAA0MV0KMQcP35c3/jGN7Rs2TL17NnTabdtW4sXL9a8efM0fvx4paWladWqVTp58qTWrVsnSQoEAlq+fLmeffZZjRw5UrfeeqvWrFmjt99+W5s2bZIk7dq1SxUVFfrpT3+qzMxMZWZmatmyZfr1r3+t3bt3X4VhAwAA03UoxDzyyCMaN26cRo4cGdS+Z88e+f1+jRo1ymlzu90aPny4tmzZIknasWOHTp06FVTj8/mUlpbm1FRXV8uyLGVkZDg1w4YNk2VZTs25Wlpa1NjYGLQAAICuKzzUDdavX68333xT27dvb7fO7/dLkjweT1C7x+PRvn37nJrIyMigGZyzNWe39/v9SkpKarf/pKQkp+ZcpaWleuqpp0IdDgAAMFRIMzEHDhzQt771La1Zs0Y9evS4YJ3L5Qp6bdt2u7ZznVtzvvqL7Wfu3LkKBALOcuDAgYseDwAAmC2kELNjxw7V19crPT1d4eHhCg8P1+bNm/XDH/5Q4eHhzgzMubMl9fX1zjqv16vW1lY1NDRctObIkSPtjn/06NF2szxnud1uxcXFBS0AAKDrCinEZGdn6+2331Ztba2zDB06VN/4xjdUW1urT3/60/J6vaqsrHS2aW1t1ebNm5WVlSVJSk9PV0RERFDN4cOHVVdX59RkZmYqEAho27ZtTs3WrVsVCAScGgAA0L2FdE1MbGys0tLSgtpiYmKUkJDgtBcWFqqkpESpqalKTU1VSUmJoqOjlZ+fL0myLEuTJk3SrFmzlJCQoPj4eBUVFWnw4MHOhcIDBw7UmDFjNHnyZC1dulSSNGXKFOXm5mrAgAFXPGgAAGC+kC/svZTZs2erublZ06dPV0NDgzIyMrRx40bFxsY6NYsWLVJ4eLgmTJig5uZmZWdna+XKlQoLC3Nq1q5dq5kzZzp3MeXl5amsrOxqdxcAABjKZdu23dmd+CQ0NjbKsiwFAoFud31Mv8c3dHYXcA3tfWZcZ3cB1xDnd/fSHc/vUD6/+dtJAADASIQYAABgJEIMAAAwEiEGAAAYiRADAACMRIgBAABGIsQAAAAjEWIAAICRCDEAAMBIhBgAAGAkQgwAADASIQYAABiJEAMAAIxEiAEAAEYixAAAACMRYgAAgJEIMQAAwEiEGAAAYCRCDAAAMBIhBgAAGIkQAwAAjESIAQAARiLEAAAAIxFiAACAkQgxAADASIQYAABgJEIMAAAwEiEGAAAYiRADAACMRIgBAABGIsQAAAAjEWIAAICRQgoxS5Ys0ZAhQxQXF6e4uDhlZmbqN7/5jbPetm0VFxfL5/MpKipKI0aM0M6dO4P20dLSohkzZigxMVExMTHKy8vTwYMHg2oaGhpUUFAgy7JkWZYKCgp07Nixjo8SAAB0OSGFmD59+uiZZ57RG2+8oTfeeEN33323vvzlLztBZcGCBVq4cKHKysq0fft2eb1e5eTkqKmpydlHYWGhysvLtX79elVVVen48ePKzc1VW1ubU5Ofn6/a2lpVVFSooqJCtbW1KigouEpDBgAAXYHLtm37SnYQHx+v733ve3rooYfk8/lUWFioOXPmSPpo1sXj8Wj+/PmaOnWqAoGAevXqpdWrV+v++++XJB06dEjJycl69dVXNXr0aO3atUuDBg1STU2NMjIyJEk1NTXKzMzUu+++qwEDBlxWvxobG2VZlgKBgOLi4q5kiMbp9/iGzu4CrqG9z4zr7C7gGuL87l664/kdyud3h6+JaWtr0/r163XixAllZmZqz5498vv9GjVqlFPjdrs1fPhwbdmyRZK0Y8cOnTp1KqjG5/MpLS3NqamurpZlWU6AkaRhw4bJsiyn5nxaWlrU2NgYtAAAgK4r5BDz9ttv61Of+pTcbremTZum8vJyDRo0SH6/X5Lk8XiC6j0ej7PO7/crMjJSPXv2vGhNUlJSu+MmJSU5NedTWlrqXENjWZaSk5NDHRoAADBIyCFmwIABqq2tVU1Njf7jP/5DEydO1DvvvOOsd7lcQfW2bbdrO9e5Neerv9R+5s6dq0Ag4CwHDhy43CEBAAADhRxiIiMj9ZnPfEZDhw5VaWmpbrnlFv3gBz+Q1+uVpHazJfX19c7sjNfrVWtrqxoaGi5ac+TIkXbHPXr0aLtZno9zu93OXVNnFwAA0HVd8XNibNtWS0uLUlJS5PV6VVlZ6axrbW3V5s2blZWVJUlKT09XREREUM3hw4dVV1fn1GRmZioQCGjbtm1OzdatWxUIBJwaAACA8FCKv/3tb2vs2LFKTk5WU1OT1q9frz/84Q+qqKiQy+VSYWGhSkpKlJqaqtTUVJWUlCg6Olr5+fmSJMuyNGnSJM2aNUsJCQmKj49XUVGRBg8erJEjR0qSBg4cqDFjxmjy5MlaunSpJGnKlCnKzc297DuTAABA1xdSiDly5IgKCgp0+PBhWZalIUOGqKKiQjk5OZKk2bNnq7m5WdOnT1dDQ4MyMjK0ceNGxcbGOvtYtGiRwsPDNWHCBDU3Nys7O1srV65UWFiYU7N27VrNnDnTuYspLy9PZWVlV2O8AACgi7ji58Rcr3hODLqL7vgcie6M87t76Y7n9zV5TgwAAEBnIsQAAAAjEWIAAICRCDEAAMBIhBgAAGAkQgwAADASIQYAABiJEAMAAIxEiAEAAEYixAAAACMRYgAAgJEIMQAAwEiEGAAAYCRCDAAAMBIhBgAAGIkQAwAAjESIAQAARiLEAAAAIxFiAACAkQgxAADASIQYAABgJEIMAAAwEiEGAAAYiRADAACMRIgBAABGIsQAAAAjEWIAAICRCDEAAMBIhBgAAGAkQgwAADASIQYAABiJEAMAAIxEiAEAAEYKKcSUlpbqtttuU2xsrJKSkvSVr3xFu3fvDqqxbVvFxcXy+XyKiorSiBEjtHPnzqCalpYWzZgxQ4mJiYqJiVFeXp4OHjwYVNPQ0KCCggJZliXLslRQUKBjx451bJQAAKDLCSnEbN68WY888ohqampUWVmp06dPa9SoUTpx4oRTs2DBAi1cuFBlZWXavn27vF6vcnJy1NTU5NQUFhaqvLxc69evV1VVlY4fP67c3Fy1tbU5Nfn5+aqtrVVFRYUqKipUW1urgoKCqzBkAADQFbhs27Y7uvHRo0eVlJSkzZs364tf/KJs25bP51NhYaHmzJkj6aNZF4/Ho/nz52vq1KkKBALq1auXVq9erfvvv1+SdOjQISUnJ+vVV1/V6NGjtWvXLg0aNEg1NTXKyMiQJNXU1CgzM1PvvvuuBgwYcMm+NTY2yrIsBQIBxcXFdXSIRur3+IbO7gKuob3PjOvsLuAa4vzuXrrj+R3K5/cVXRMTCAQkSfHx8ZKkPXv2yO/3a9SoUU6N2+3W8OHDtWXLFknSjh07dOrUqaAan8+ntLQ0p6a6ulqWZTkBRpKGDRsmy7KcmnO1tLSosbExaAEAAF1Xh0OMbdt67LHHdMcddygtLU2S5Pf7JUkejyeo1uPxOOv8fr8iIyPVs2fPi9YkJSW1O2ZSUpJTc67S0lLn+hnLspScnNzRoQEAAAN0OMR885vf1J///Ge9+OKL7da5XK6g17Ztt2s717k156u/2H7mzp2rQCDgLAcOHLicYQAAAEN1KMTMmDFDr7zyin7/+9+rT58+TrvX65WkdrMl9fX1zuyM1+tVa2urGhoaLlpz5MiRdsc9evRou1mes9xut+Li4oIWAADQdYUUYmzb1je/+U29/PLLeu2115SSkhK0PiUlRV6vV5WVlU5ba2urNm/erKysLElSenq6IiIigmoOHz6suro6pyYzM1OBQEDbtm1zarZu3apAIODUAACA7i08lOJHHnlE69at0y9/+UvFxsY6My6WZSkqKkoul0uFhYUqKSlRamqqUlNTVVJSoujoaOXn5zu1kyZN0qxZs5SQkKD4+HgVFRVp8ODBGjlypCRp4MCBGjNmjCZPnqylS5dKkqZMmaLc3NzLujMJAAB0fSGFmCVLlkiSRowYEdS+YsUKPfjgg5Kk2bNnq7m5WdOnT1dDQ4MyMjK0ceNGxcbGOvWLFi1SeHi4JkyYoObmZmVnZ2vlypUKCwtzatauXauZM2c6dzHl5eWprKysI2MEAABd0BU9J+Z6xnNi0F10x+dIdGec391Ldzy/r9lzYgAAADoLIQYAABiJEAMAAIxEiAEAAEYixAAAACMRYgAAgJEIMQAAwEiEGAAAYCRCDAAAMBIhBgAAGIkQAwAAjESIAQAARiLEAAAAIxFiAACAkQgxAADASIQYAABgJEIMAAAwEiEGAAAYiRADAACMRIgBAABGIsQAAAAjEWIAAICRCDEAAMBIhBgAAGAkQgwAADASIQYAABiJEAMAAIxEiAEAAEYixAAAACMRYgAAgJEIMQAAwEiEGAAAYKSQQ8wf//hH3XPPPfL5fHK5XPrFL34RtN62bRUXF8vn8ykqKkojRozQzp07g2paWlo0Y8YMJSYmKiYmRnl5eTp48GBQTUNDgwoKCmRZlizLUkFBgY4dOxbyAAEAQNcUcog5ceKEbrnlFpWVlZ13/YIFC7Rw4UKVlZVp+/bt8nq9ysnJUVNTk1NTWFio8vJyrV+/XlVVVTp+/Lhyc3PV1tbm1OTn56u2tlYVFRWqqKhQbW2tCgoKOjBEAADQFYWHusHYsWM1duzY866zbVuLFy/WvHnzNH78eEnSqlWr5PF4tG7dOk2dOlWBQEDLly/X6tWrNXLkSEnSmjVrlJycrE2bNmn06NHatWuXKioqVFNTo4yMDEnSsmXLlJmZqd27d2vAgAEdHS8AAOgiruo1MXv27JHf79eoUaOcNrfbreHDh2vLli2SpB07dujUqVNBNT6fT2lpaU5NdXW1LMtyAowkDRs2TJZlOTXnamlpUWNjY9ACAAC6rqsaYvx+vyTJ4/EEtXs8Hmed3+9XZGSkevbsedGapKSkdvtPSkpyas5VWlrqXD9jWZaSk5OveDwAAOD69YncneRyuYJe27bdru1c59acr/5i+5k7d64CgYCzHDhwoAM9BwAApriqIcbr9UpSu9mS+vp6Z3bG6/WqtbVVDQ0NF605cuRIu/0fPXq03SzPWW63W3FxcUELAADouq5qiElJSZHX61VlZaXT1traqs2bNysrK0uSlJ6eroiIiKCaw4cPq66uzqnJzMxUIBDQtm3bnJqtW7cqEAg4NQAAoHsL+e6k48eP669//avzes+ePaqtrVV8fLxuuukmFRYWqqSkRKmpqUpNTVVJSYmio6OVn58vSbIsS5MmTdKsWbOUkJCg+Ph4FRUVafDgwc7dSgMHDtSYMWM0efJkLV26VJI0ZcoU5ebmcmcSAACQ1IEQ88Ybb+iuu+5yXj/22GOSpIkTJ2rlypWaPXu2mpubNX36dDU0NCgjI0MbN25UbGyss82iRYsUHh6uCRMmqLm5WdnZ2Vq5cqXCwsKcmrVr12rmzJnOXUx5eXkXfDYNAADofly2bdud3YlPQmNjoyzLUiAQ6HbXx/R7fENndwHX0N5nxnV2F3ANcX53L93x/A7l85u/nQQAAIxEiAEAAEYixAAAACMRYgAAgJEIMQAAwEiEGAAAYCRCDAAAMBIhBgAAGIkQAwAAjESIAQAARiLEAAAAIxFiAACAkQgxAADASIQYAABgJEIMAAAwEiEGAAAYiRADAACMRIgBAABGIsQAAAAjEWIAAICRCDEAAMBIhBgAAGAkQgwAADASIQYAABiJEAMAAIxEiAEAAEYixAAAACMRYgAAgJEIMQAAwEiEGAAAYCRCDAAAMBIhBgAAGOm6DzHPPfecUlJS1KNHD6Wnp+v111/v7C4BAIDrwHUdYl566SUVFhZq3rx5+tOf/qQ777xTY8eO1f79+zu7awAAoJNd1yFm4cKFmjRpkh5++GENHDhQixcvVnJyspYsWdLZXQMAAJ0svLM7cCGtra3asWOHHn/88aD2UaNGacuWLe3qW1pa1NLS4rwOBAKSpMbGxk+2o9ehMy0nO7sLuIa64+94d8b53b10x/P77Jht275k7XUbYj744AO1tbXJ4/EEtXs8Hvn9/nb1paWleuqpp9q1Jycnf2J9BK4H1uLO7gGAT0p3Pr+bmppkWdZFa67bEHOWy+UKem3bdrs2SZo7d64ee+wx5/WZM2f0j3/8QwkJCeetR9fS2Nio5ORkHThwQHFxcZ3dHQBXEed392LbtpqamuTz+S5Ze92GmMTERIWFhbWbdamvr283OyNJbrdbbrc7qO3GG2/8JLuI61BcXBz/kwO6KM7v7uNSMzBnXbcX9kZGRio9PV2VlZVB7ZWVlcrKyuqkXgEAgOvFdTsTI0mPPfaYCgoKNHToUGVmZur555/X/v37NW3atM7uGgAA6GTXdYi5//779eGHH+rpp5/W4cOHlZaWpldffVV9+/bt7K7hOuN2u/Xkk0+2+0oRgPk4v3EhLvty7mECAAC4zly318QAAABcDCEGAAAYiRADAACMRIgBAABGIsQAAAAjEWIAAICRruvnxAAXcvDgQS1ZskRbtmyR3++Xy+WSx+NRVlaWpk2bxh/+BIBugOfEwDhVVVUaO3askpOTNWrUKHk8Htm2rfr6elVWVurAgQP6zW9+o9tvv72zuwrgE3DgwAE9+eST+tnPftbZXUEnI8TAOLfddpvuuOMOLVq06LzrH330UVVVVWn79u3XuGcAroW33npLX/jCF9TW1tbZXUEnI8TAOFFRUaqtrdWAAQPOu/7dd9/Vrbfequbm5mvcMwBXwyuvvHLR9e+//75mzZpFiAHXxMA8vXv31pYtWy4YYqqrq9W7d+9r3CsAV8tXvvIVuVwuXezf2C6X6xr2CNcrQgyMU1RUpGnTpmnHjh3KycmRx+ORy+WS3+9XZWWlfvrTn2rx4sWd3U0AHdS7d2/9+Mc/1le+8pXzrq+trVV6evq17RSuS4QYGGf69OlKSEjQokWLtHTpUmdKOSwsTOnp6XrhhRc0YcKETu4lgI5KT0/Xm2++ecEQc6lZGnQfXBMDo506dUoffPCBJCkxMVERERGd3CMAV+r111/XiRMnNGbMmPOuP3HihN544w0NHz78GvcM1xtCDAAAMBJP7AUAAEYixAAAACMRYgAAgJEIMQAAwEiEGAAAYCRCDAAAMBIhBgAAGOn/AACPSLo4noLMAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Distribution of label (Target var))\n",
    "h4.label.value_counts().sort_index().plot.bar()\n",
    "plt.title('Distribution of Target')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faa5855a",
   "metadata": {},
   "source": [
    "# Preparation\n",
    "- Clean IDs and create useful features such as nucleotide distributions\n",
    "- Check lengths of sequences for any unusual lengths (!= 500)\n",
    "- Vectorize sequences into hexamers and codons via Bag of Words CountVectorizer\n",
    "- Train, Validate, Test split:\n",
    "    - Split X data for codons and hexamers\n",
    "    - Split y data for labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7471dc7",
   "metadata": {},
   "source": [
    "### Cleaning data and checking for discrepancies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ba288340",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>iTELL-Chr1_61</td>\n",
       "      <td>CACACCACACCCACACACCCACACACCACACCACACACCACACCAC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>iYAL067W-A_3174</td>\n",
       "      <td>TTTCTAAGTGGGATTTTTCTTAATCCTTGGATTCTTAAAAGGTTAT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>iYAL067W-A_4141</td>\n",
       "      <td>AAAGCCATTAAGGTCCAGATAACCAAGGGACAATAAACCTATGCTT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>iYAL067W-A_6428</td>\n",
       "      <td>CCATCATTTTACTTCCACATAAGTTCTAATGTTACGTAGTTCAATT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>iYAL067W-A_6786</td>\n",
       "      <td>AATTATATTTCCATCAGCTCAATACCGCAGTACTTTGAAACCTGAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                id                                           sequence  label\n",
       "0    iTELL-Chr1_61  CACACCACACCCACACACCCACACACCACACCACACACCACACCAC...      0\n",
       "1  iYAL067W-A_3174  TTTCTAAGTGGGATTTTTCTTAATCCTTGGATTCTTAAAAGGTTAT...      1\n",
       "2  iYAL067W-A_4141  AAAGCCATTAAGGTCCAGATAACCAAGGGACAATAAACCTATGCTT...      1\n",
       "3  iYAL067W-A_6428  CCATCATTTTACTTCCACATAAGTTCTAATGTTACGTAGTTCAATT...      0\n",
       "4  iYAL067W-A_6786  AATTATATTTCCATCAGCTCAATACCGCAGTACTTTGAAACCTGAT...      0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cleaning id strings with \\n appending\n",
    "h4.id = h4.id.apply(lambda x: x.replace('\\n', '')) # deleting /n appending each id\n",
    "h4.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6a51030d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14601, 3)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h4.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fc6e74c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unexpected sequence length at index: 0, length 310.\n",
      "Unexpected sequence length at index: 263, length 449.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>iTELL-Chr1_61</td>\n",
       "      <td>CACACCACACCCACACACCCACACACCACACCACACACCACACCAC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>iTELR-Chr1_230009</td>\n",
       "      <td>TTATATTTGTTTAAATTGGATTTGTTTACATTAGATTTGTTTACAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id                                           sequence  \\\n",
       "0        iTELL-Chr1_61  CACACCACACCCACACACCCACACACCACACCACACACCACACCAC...   \n",
       "263  iTELR-Chr1_230009  TTATATTTGTTTAAATTGGATTTGTTTACATTAGATTTGTTTACAT...   \n",
       "\n",
       "    label  \n",
       "0       0  \n",
       "263     0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking sequence lengths for abnormal (!= 500) lengths\n",
    "unexpected = pr.check_len(h4)\n",
    "unexpected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "97fe8a4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dropping unexpected rows\n",
    "h4 = h4.drop(unexpected.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "36e24112",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, sequence, label]\n",
       "Index: []"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pr.check_len(h4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c55c10de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>iYAL067W-A_3174</td>\n",
       "      <td>TTTCTAAGTGGGATTTTTCTTAATCCTTGGATTCTTAAAAGGTTAT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>iYAL067W-A_4141</td>\n",
       "      <td>AAAGCCATTAAGGTCCAGATAACCAAGGGACAATAAACCTATGCTT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>iYAL067W-A_6428</td>\n",
       "      <td>CCATCATTTTACTTCCACATAAGTTCTAATGTTACGTAGTTCAATT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>iYAL067W-A_6786</td>\n",
       "      <td>AATTATATTTCCATCAGCTCAATACCGCAGTACTTTGAAACCTGAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>iYAL067W-A_7060</td>\n",
       "      <td>AACAATAGTGGGTATGAGTAAAGATATATAGATCGATATTTTGAAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                id                                           sequence  label\n",
       "0  iYAL067W-A_3174  TTTCTAAGTGGGATTTTTCTTAATCCTTGGATTCTTAAAAGGTTAT...      1\n",
       "1  iYAL067W-A_4141  AAAGCCATTAAGGTCCAGATAACCAAGGGACAATAAACCTATGCTT...      1\n",
       "2  iYAL067W-A_6428  CCATCATTTTACTTCCACATAAGTTCTAATGTTACGTAGTTCAATT...      0\n",
       "3  iYAL067W-A_6786  AATTATATTTCCATCAGCTCAATACCGCAGTACTTTGAAACCTGAT...      0\n",
       "4  iYAL067W-A_7060  AACAATAGTGGGTATGAGTAAAGATATATAGATCGATATTTTGAAT...      0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h4 = h4.reset_index(drop = True)\n",
    "h4.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e71789cb",
   "metadata": {},
   "source": [
    "### Cleaning takeaways:\n",
    "- IDs had appending /n char\n",
    "- Two entries had a sequence length != 500\n",
    "    - The two entries were both chromosome sequences, and both were label 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ea0d9b6",
   "metadata": {},
   "source": [
    "### Pre-exploration Feature Creation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3d751182",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>label</th>\n",
       "      <th>adenine</th>\n",
       "      <th>cytosine</th>\n",
       "      <th>guanine</th>\n",
       "      <th>thymine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>iYAL067W-A_3174</td>\n",
       "      <td>TTTCTAAGTGGGATTTTTCTTAATCCTTGGATTCTTAAAAGGTTAT...</td>\n",
       "      <td>1</td>\n",
       "      <td>168</td>\n",
       "      <td>94</td>\n",
       "      <td>76</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>iYAL067W-A_4141</td>\n",
       "      <td>AAAGCCATTAAGGTCCAGATAACCAAGGGACAATAAACCTATGCTT...</td>\n",
       "      <td>1</td>\n",
       "      <td>166</td>\n",
       "      <td>89</td>\n",
       "      <td>85</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>iYAL067W-A_6428</td>\n",
       "      <td>CCATCATTTTACTTCCACATAAGTTCTAATGTTACGTAGTTCAATT...</td>\n",
       "      <td>0</td>\n",
       "      <td>160</td>\n",
       "      <td>92</td>\n",
       "      <td>68</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>iYAL067W-A_6786</td>\n",
       "      <td>AATTATATTTCCATCAGCTCAATACCGCAGTACTTTGAAACCTGAT...</td>\n",
       "      <td>0</td>\n",
       "      <td>185</td>\n",
       "      <td>70</td>\n",
       "      <td>75</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>iYAL067W-A_7060</td>\n",
       "      <td>AACAATAGTGGGTATGAGTAAAGATATATAGATCGATATTTTGAAT...</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>61</td>\n",
       "      <td>85</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                id                                           sequence  label  \\\n",
       "0  iYAL067W-A_3174  TTTCTAAGTGGGATTTTTCTTAATCCTTGGATTCTTAAAAGGTTAT...      1   \n",
       "1  iYAL067W-A_4141  AAAGCCATTAAGGTCCAGATAACCAAGGGACAATAAACCTATGCTT...      1   \n",
       "2  iYAL067W-A_6428  CCATCATTTTACTTCCACATAAGTTCTAATGTTACGTAGTTCAATT...      0   \n",
       "3  iYAL067W-A_6786  AATTATATTTCCATCAGCTCAATACCGCAGTACTTTGAAACCTGAT...      0   \n",
       "4  iYAL067W-A_7060  AACAATAGTGGGTATGAGTAAAGATATATAGATCGATATTTTGAAT...      0   \n",
       "\n",
       "   adenine  cytosine  guanine  thymine  \n",
       "0      168        94       76      162  \n",
       "1      166        89       85      160  \n",
       "2      160        92       68      180  \n",
       "3      185        70       75      170  \n",
       "4      168        61       85      186  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Creates features for the nucleotide count of each sequence\n",
    "h4 = pr.nucleotide_count(h4)\n",
    "h4.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cecc6da3",
   "metadata": {},
   "source": [
    "## NLP Bag of Words CountVectorizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "964169de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create overlapping hexamers(6 bases) and codons(3 bases) of each sequence for processing\n",
    "h4['hexamers'] = h4.apply(lambda x: pr.getKmers(x['sequence']), axis=1)\n",
    "h4['codons'] = h4.apply(lambda x: pr.getKmers(x['sequence'], k = 3), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ff66a8b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>label</th>\n",
       "      <th>adenine</th>\n",
       "      <th>cytosine</th>\n",
       "      <th>guanine</th>\n",
       "      <th>thymine</th>\n",
       "      <th>hexamers</th>\n",
       "      <th>codons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>iYAL067W-A_3174</td>\n",
       "      <td>TTTCTAAGTGGGATTTTTCTTAATCCTTGGATTCTTAAAAGGTTAT...</td>\n",
       "      <td>1</td>\n",
       "      <td>168</td>\n",
       "      <td>94</td>\n",
       "      <td>76</td>\n",
       "      <td>162</td>\n",
       "      <td>[tttcta, ttctaa, tctaag, ctaagt, taagtg, aagtg...</td>\n",
       "      <td>[ttt, ttc, tct, cta, taa, aag, agt, gtg, tgg, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                id                                           sequence  label  \\\n",
       "0  iYAL067W-A_3174  TTTCTAAGTGGGATTTTTCTTAATCCTTGGATTCTTAAAAGGTTAT...      1   \n",
       "\n",
       "   adenine  cytosine  guanine  thymine  \\\n",
       "0      168        94       76      162   \n",
       "\n",
       "                                            hexamers  \\\n",
       "0  [tttcta, ttctaa, tctaag, ctaagt, taagtg, aagtg...   \n",
       "\n",
       "                                              codons  \n",
       "0  [ttt, ttc, tct, cta, taa, aag, agt, gtg, tgg, ...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h4.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a313bfb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating dataframes of hexamers and codons for processing by Bag of Words CountVectorizer\n",
    "hexamers = h4[['hexamers', 'label']]\n",
    "codons = h4[['codons', 'label']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "cd4815b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>codons</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[ttt, ttc, tct, cta, taa, aag, agt, gtg, tgg, ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[aaa, aag, agc, gcc, cca, cat, att, tta, taa, ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[cca, cat, atc, tca, cat, att, ttt, ttt, tta, ...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              codons  label\n",
       "0  [ttt, ttc, tct, cta, taa, aag, agt, gtg, tgg, ...      1\n",
       "1  [aaa, aag, agc, gcc, cca, cat, att, tta, taa, ...      1\n",
       "2  [cca, cat, atc, tca, cat, att, ttt, ttt, tta, ...      0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking new codon df\n",
    "codons.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "47d50828",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hexamers</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[tttcta, ttctaa, tctaag, ctaagt, taagtg, aagtg...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[aaagcc, aagcca, agccat, gccatt, ccatta, catta...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[ccatca, catcat, atcatt, tcattt, catttt, atttt...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            hexamers  label\n",
       "0  [tttcta, ttctaa, tctaag, ctaagt, taagtg, aagtg...      1\n",
       "1  [aaagcc, aagcca, agccat, gccatt, ccatta, catta...      1\n",
       "2  [ccatca, catcat, atcatt, tcattt, catttt, atttt...      0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking new hexamer df\n",
    "hexamers.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4b464247",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Joining each list of hexamers into a 'sentence' for BoW processing\n",
    "hexamer_vectors = list(hexamers['hexamers'].values)\n",
    "for i in range(len(hexamer_vectors)):\n",
    "    hexamer_vectors[i] = ' '.join(hexamer_vectors[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7d8e7bc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_vectors = list(codons['codons'].values)\n",
    "for i in range(len(codon_vectors)):\n",
    "    codon_vectors[i] = ' '.join(codon_vectors[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8d1de31b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'tttcta ttctaa tctaag ctaagt taagtg aagtgg agtggg gtggga tgggat gggatt ggattt gatttt attttt tttttc ttttct tttctt ttctta tcttaa cttaat ttaatc taatcc aatcct atcctt tccttg ccttgg cttgga ttggat tggatt ggattc gattct attctt ttctta tcttaa cttaaa ttaaaa taaaag aaaagg aaaggt aaggtt aggtta ggttat gttatt ttatta tattaa attaaa ttaaag taaagt aaagtt aagttc agttcc gttccg ttccgc tccgca ccgcac cgcaca gcacaa cacaaa acaaag caaaga aaagaa aagaac agaacg gaacgc aacgct acgctt cgcttg gcttgg cttgga ttggaa tggaaa ggaaat gaaatc aaatcg aatcgc atcgca tcgcat cgcatt gcattc cattca attcat ttcatc tcatca catcaa atcaaa tcaaag caaaga aaagaa aagaac agaaca gaacaa aacaac acaact caactc aactct actctt ctcttc tcttcg cttcgt ttcgtt tcgttt cgtttt gttttc ttttcc tttcca ttccaa tccaaa ccaaac caaaca aaacaa aacaat acaatc caatct aatctt atcttc tcttcc cttccc ttcccg tcccga cccgaa ccgaaa cgaaaa gaaaaa aaaaag aaaagt aaagta aagtag agtagc gtagcc tagccg agccgt gccgtt ccgttc cgttca gttcat ttcatt tcattt catttc atttcc tttccc ttccct tccctt cccttc ccttcc cttccg ttccga tccgat ccgatt cgattt gatttc atttca tttcat ttcatt tcattc cattcc attcct ttccta tcctag cctaga ctagac tagact agactg gactgc actgcc ctgcca tgccaa gccaaa ccaaat caaatt aaattt aatttt attttt tttttc ttttct tttctt ttcttg tcttgc cttgct ttgctc tgctca gctcat ctcatt tcattt cattta atttat tttata ttataa tataat ataatg taatga aatgat atgatt tgattg gattga attgat ttgata tgataa gataag ataaga taagaa aagaat agaatt gaattg aattgt attgta ttgtat tgtatt gtattt tatttg atttgt tttgtg ttgtgt tgtgtc gtgtcc tgtccc gtccca tcccat cccatt ccattc cattct attctc ttctcg tctcgt ctcgta tcgtag cgtaga gtagat tagata agataa gataaa ataaaa taaaat aaaatt aaattc aattct attctt ttcttg tcttgg cttgga ttggat tggatg ggatgt gatgtt atgtta tgttaa gttaaa ttaaaa taaaaa aaaaaa aaaaat aaaatt aaatta aattat attatt ttatta tattat attatt ttattt tatttt attttc ttttct tttctt ttcttc tcttca cttcat ttcata tcataa cataaa ataaag taaaga aaagaa aagaag agaagc gaagct aagctt agcttt gctttc ctttca tttcaa ttcaag tcaaga caagat aagata agatat gatata atataa tataag ataaga taagat aagata agatac gatacg atacga tacgaa acgaaa cgaaat gaaata aaatag aatagg ataggg tagggg aggggt ggggtt gggttg ggttga gttgat ttgata tgataa gataat ataatt taattg aattgc attgca ttgcat tgcatg gcatga catgac atgaca tgacag gacagt acagta cagtag agtagc gtagct tagctt agcttt gcttta ctttag tttaga ttagat tagatc agatca gatcaa atcaaa tcaaaa caaaaa aaaaag aaaagg aaagga aaggaa aggaaa ggaaag gaaagc aaagca aagcat agcatg gcatgg catgga atggag tggagg ggaggg gaggga agggaa gggaaa ggaaac gaaaca aaacag aacagt acagta cagtaa agtaaa gtaaac taaaca aaacag aacagt acagtg cagtga agtgaa gtgaaa tgaaaa gaaaat aaaatt aaattc aattct attctc ttctct tctctt ctcttg tcttga cttgag ttgaga tgagaa gagaac agaacc gaacca aaccaa accaaa ccaaag caaagt aaagta aagtaa agtaaa gtaaac taaacc aaacct aacctt accttc ccttca cttcat ttcatt tcattg cattga attgaa ttgaag tgaaga gaagag aagagc agagct gagctt agcttc gcttcc cttcct ttcctt tcctta ccttaa cttaaa ttaaaa taaaaa aaaaaa aaaaat aaaatt aaattt aattta atttag tttaga ttagaa tagaat agaatc gaatct aatctc atctcc tctccc ctccca tcccat cccatg ccatgt catgtc atgtca tgtcaa gtcaac tcaacg caacgg aacggg acgggt cgggtt gggttt ggtttc gtttcc tttcca ttccat tccata ccatac catacc atacct tacctc acctcc cctccc ctcccc tcccca ccccag cccagc ccagca cagcat agcatc gcatca catcat atcata tcatac cataca atacat tacatc acatct catctt atcttt tctttt cttttt tttttt tttttc ttttca tttcaa ttcaaa tcaaag caaaga aaagaa aagaaa agaaac gaaact aaactt aacttc acttca cttcaa ttcaaa tcaaat caaatg aaatgc aatgcc'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hexamer_vectors[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8f9b462a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ttt ttc tct cta taa aag agt gtg tgg ggg gga gat att ttt ttt ttt ttc tct ctt tta taa aat atc tcc cct ctt ttg tgg gga gat att ttc tct ctt tta taa aaa aaa aag agg ggt gtt tta tat att tta taa aaa aag agt gtt ttc tcc ccg cgc gca cac aca caa aaa aag aga gaa aac acg cgc gct ctt ttg tgg gga gaa aaa aat atc tcg cgc gca cat att ttc tca cat atc tca caa aaa aag aga gaa aac aca caa aac act ctc tct ctt ttc tcg cgt gtt ttt ttt ttc tcc cca caa aaa aac aca caa aat atc tct ctt ttc tcc ccc ccg cga gaa aaa aaa aaa aag agt gta tag agc gcc ccg cgt gtt ttc tca cat att ttt ttc tcc ccc cct ctt ttc tcc ccg cga gat att ttt ttc tca cat att ttc tcc cct cta tag aga gac act ctg tgc gcc cca caa aaa aat att ttt ttt ttt ttc tct ctt ttg tgc gct ctc tca cat att ttt tta tat ata taa aat atg tga gat att ttg tga gat ata taa aag aga gaa aat att ttg tgt gta tat att ttt ttg tgt gtg tgt gtc tcc ccc cca cat att ttc tct ctc tcg cgt gta tag aga gat ata taa aaa aaa aat att ttc tct ctt ttg tgg gga gat atg tgt gtt tta taa aaa aaa aaa aaa aat att tta tat att tta tat att ttt ttt ttc tct ctt ttc tca cat ata taa aaa aag aga gaa aag agc gct ctt ttt ttc tca caa aag aga gat ata tat ata taa aag aga gat ata tac acg cga gaa aaa aat ata tag agg ggg ggg ggt gtt ttg tga gat ata taa aat att ttg tgc gca cat atg tga gac aca cag agt gta tag agc gct ctt ttt tta tag aga gat atc tca caa aaa aaa aaa aag agg gga gaa aaa aag agc gca cat atg tgg gga gag agg ggg gga gaa aaa aac aca cag agt gta taa aaa aac aca cag agt gtg tga gaa aaa aaa aat att ttc tct ctc tct ctt ttg tga gag aga gaa aac acc cca caa aaa aag agt gta taa aaa aac acc cct ctt ttc tca cat att ttg tga gaa aag aga gag agc gct ctt ttc tcc cct ctt tta taa aaa aaa aaa aaa aat att ttt tta tag aga gaa aat atc tct ctc tcc ccc cca cat atg tgt gtc tca caa aac acg cgg ggg ggt gtt ttt ttc tcc cca cat ata tac acc cct ctc tcc ccc ccc cca cag agc gca cat atc tca cat ata tac aca cat atc tct ctt ttt ttt ttt ttt ttc tca caa aaa aag aga gaa aaa aac act ctt ttc tca caa aaa aat atg tgc gcc'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_vectors[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "794a066b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Using Bag of Words count vectorizer for hexamers\n",
    "cv = CountVectorizer(ngram_range=(1,1))\n",
    "X_hex = cv.fit_transform(hexamer_vectors).toarray()\n",
    "unique_hexamers = cv.get_feature_names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "27abdc9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using Bag of Words count vectorizer for codons\n",
    "X_cod = cv.fit_transform(codon_vectors).toarray()\n",
    "unique_codons = cv.get_feature_names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fa2eac95",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([2, 0, 2, ..., 3, 0, 1]),\n",
       " array([36, 10, 16, 14,  7,  3,  3,  3, 13,  6,  4,  7, 11,  8,  6, 21, 12,\n",
       "         1,  4, 15,  7,  6,  4,  6,  3,  3,  1,  3,  2,  6,  1, 18, 14,  2,\n",
       "         3, 11,  5,  3,  0,  5,  7,  0,  5,  3,  6,  2,  3,  7, 14,  3,  7,\n",
       "         6, 13, 12,  3, 13,  7,  4,  5,  5, 11, 25, 11, 23]))"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking data\n",
    "X_hex[0], X_cod[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "787afcf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating DFs from array of arrays for hexamers and codons\n",
    "X_hex = pd.DataFrame(X_hex, columns = unique_hexamers)\n",
    "X_cod = pd.DataFrame(X_cod, columns = unique_codons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f5438f6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding ACGT counts to hexamers and codons\n",
    "X_hex['adenine'] = h4.adenine\n",
    "X_hex['cytosine'] = h4.cytosine\n",
    "X_hex['guanine'] = h4.guanine\n",
    "X_hex['thymine'] = h4.thymine\n",
    "\n",
    "X_cod['adenine'] = h4.adenine\n",
    "X_cod['cytosine'] = h4.cytosine\n",
    "X_cod['guanine'] = h4.guanine\n",
    "X_cod['thymine'] = h4.thymine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b260e66f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aaaaaa</th>\n",
       "      <th>aaaaac</th>\n",
       "      <th>aaaaag</th>\n",
       "      <th>aaaaat</th>\n",
       "      <th>aaaaca</th>\n",
       "      <th>aaaacc</th>\n",
       "      <th>aaaacg</th>\n",
       "      <th>aaaact</th>\n",
       "      <th>aaaaga</th>\n",
       "      <th>aaaagc</th>\n",
       "      <th>...</th>\n",
       "      <th>ttttgg</th>\n",
       "      <th>ttttgt</th>\n",
       "      <th>ttttta</th>\n",
       "      <th>tttttc</th>\n",
       "      <th>tttttg</th>\n",
       "      <th>tttttt</th>\n",
       "      <th>adenine</th>\n",
       "      <th>cytosine</th>\n",
       "      <th>guanine</th>\n",
       "      <th>thymine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>168</td>\n",
       "      <td>94</td>\n",
       "      <td>76</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>166</td>\n",
       "      <td>89</td>\n",
       "      <td>85</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>160</td>\n",
       "      <td>92</td>\n",
       "      <td>68</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>185</td>\n",
       "      <td>70</td>\n",
       "      <td>75</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>168</td>\n",
       "      <td>61</td>\n",
       "      <td>85</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  4100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   aaaaaa  aaaaac  aaaaag  aaaaat  aaaaca  aaaacc  aaaacg  aaaact  aaaaga  \\\n",
       "0       2       0       2       2       0       0       0       0       0   \n",
       "1       1       0       1       2       0       0       0       0       1   \n",
       "2       1       0       0       1       0       0       0       0       0   \n",
       "3      15       0       1       2       1       0       0       0       1   \n",
       "4       0       0       2       1       0       0       0       1       1   \n",
       "\n",
       "   aaaagc  ...  ttttgg  ttttgt  ttttta  tttttc  tttttg  tttttt  adenine  \\\n",
       "0       0  ...       0       0       0       3       0       1      168   \n",
       "1       0  ...       1       1       0       0       1       2      166   \n",
       "2       0  ...       1       1       0       0       0       0      160   \n",
       "3       0  ...       2       0       2       0       0       5      185   \n",
       "4       0  ...       1       2       2       2       0       7      168   \n",
       "\n",
       "   cytosine  guanine  thymine  \n",
       "0        94       76      162  \n",
       "1        89       85      160  \n",
       "2        92       68      180  \n",
       "3        70       75      170  \n",
       "4        61       85      186  \n",
       "\n",
       "[5 rows x 4100 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_hex.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4106ddac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aaa</th>\n",
       "      <th>aac</th>\n",
       "      <th>aag</th>\n",
       "      <th>aat</th>\n",
       "      <th>aca</th>\n",
       "      <th>acc</th>\n",
       "      <th>acg</th>\n",
       "      <th>act</th>\n",
       "      <th>aga</th>\n",
       "      <th>agc</th>\n",
       "      <th>...</th>\n",
       "      <th>tgg</th>\n",
       "      <th>tgt</th>\n",
       "      <th>tta</th>\n",
       "      <th>ttc</th>\n",
       "      <th>ttg</th>\n",
       "      <th>ttt</th>\n",
       "      <th>adenine</th>\n",
       "      <th>cytosine</th>\n",
       "      <th>guanine</th>\n",
       "      <th>thymine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>25</td>\n",
       "      <td>11</td>\n",
       "      <td>23</td>\n",
       "      <td>168</td>\n",
       "      <td>94</td>\n",
       "      <td>76</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>11</td>\n",
       "      <td>16</td>\n",
       "      <td>19</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>23</td>\n",
       "      <td>166</td>\n",
       "      <td>89</td>\n",
       "      <td>85</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>15</td>\n",
       "      <td>11</td>\n",
       "      <td>26</td>\n",
       "      <td>160</td>\n",
       "      <td>92</td>\n",
       "      <td>68</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>41</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>19</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "      <td>31</td>\n",
       "      <td>185</td>\n",
       "      <td>70</td>\n",
       "      <td>75</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>24</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>41</td>\n",
       "      <td>168</td>\n",
       "      <td>61</td>\n",
       "      <td>85</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   aaa  aac  aag  aat  aca  acc  acg  act  aga  agc  ...  tgg  tgt  tta  ttc  \\\n",
       "0   36   10   16   14    7    3    3    3   13    6  ...    5    5   11   25   \n",
       "1   16   11   16   19   12    5    3    8    6    9  ...    4   12   13   11   \n",
       "2   20    7   10   20    4    8    3    7    8    4  ...    4    3   26   15   \n",
       "3   41   12    9   19    8    6    1    9    5    3  ...    3    3   16    8   \n",
       "4   24    9   10   17    5    3    5    8    9    3  ...    4    9   19   11   \n",
       "\n",
       "   ttg  ttt  adenine  cytosine  guanine  thymine  \n",
       "0   11   23      168        94       76      162  \n",
       "1   10   23      166        89       85      160  \n",
       "2   11   26      160        92       68      180  \n",
       "3   14   31      185        70       75      170  \n",
       "4    9   41      168        61       85      186  \n",
       "\n",
       "[5 rows x 68 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_cod.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a6ea86c",
   "metadata": {},
   "source": [
    "### Train, Validate, Test split:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a04f3b24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1, 1, 0, 0, 0])"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_data = h4.label.values\n",
    "y_data[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9ad3008b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Splitting the data into X/y splits and exploratory splits\n",
    "Hex_train, Hex_validate, Hex_test, y_train, y_validate, y_test = pr.tvt_split(X_hex, y_data)\n",
    "\n",
    "# Splitting codons into train/test\n",
    "Cod_train, Cod_validate, Cod_test = pr.tvt_split(X_cod)\n",
    "\n",
    "# Splitting h4 for exploration\n",
    "explore_train, explore_validate, explore_test = pr.tvt_split(h4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d4b70549",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((8175, 4100), (3504, 4100), (2920, 4100))"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking hexamer subset shapes\n",
    "Hex_train.shape, Hex_validate.shape, Hex_test.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "cb39d0ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((8175, 68), (3504, 68), (2920, 68))"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking codon subset shapes\n",
    "Cod_train.shape, Cod_validate.shape, Cod_test.shape "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f94f3fa1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True])"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sanity check\n",
    "y_train[:10] == explore_train.label.values[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a554645b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aaaaaa</th>\n",
       "      <th>aaaaac</th>\n",
       "      <th>aaaaag</th>\n",
       "      <th>aaaaat</th>\n",
       "      <th>aaaaca</th>\n",
       "      <th>aaaacc</th>\n",
       "      <th>aaaacg</th>\n",
       "      <th>aaaact</th>\n",
       "      <th>aaaaga</th>\n",
       "      <th>aaaagc</th>\n",
       "      <th>...</th>\n",
       "      <th>ttttgg</th>\n",
       "      <th>ttttgt</th>\n",
       "      <th>ttttta</th>\n",
       "      <th>tttttc</th>\n",
       "      <th>tttttg</th>\n",
       "      <th>tttttt</th>\n",
       "      <th>adenine</th>\n",
       "      <th>cytosine</th>\n",
       "      <th>guanine</th>\n",
       "      <th>thymine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14085</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>164</td>\n",
       "      <td>89</td>\n",
       "      <td>104</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13466</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>157</td>\n",
       "      <td>99</td>\n",
       "      <td>103</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8961</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>160</td>\n",
       "      <td>88</td>\n",
       "      <td>88</td>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2542</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>189</td>\n",
       "      <td>102</td>\n",
       "      <td>90</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8877</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>156</td>\n",
       "      <td>85</td>\n",
       "      <td>108</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  4100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       aaaaaa  aaaaac  aaaaag  aaaaat  aaaaca  aaaacc  aaaacg  aaaact  aaaaga  \\\n",
       "14085       1       1       0       1       0       0       0       2       0   \n",
       "13466       0       0       0       0       0       0       0       0       0   \n",
       "8961        5       1       0       3       0       0       0       1       0   \n",
       "2542        0       2       0       3       3       1       0       1       1   \n",
       "8877        0       0       1       0       1       0       0       0       1   \n",
       "\n",
       "       aaaagc  ...  ttttgg  ttttgt  ttttta  tttttc  tttttg  tttttt  adenine  \\\n",
       "14085       0  ...       1       0       0       0       1       3      164   \n",
       "13466       1  ...       0       0       0       0       0       0      157   \n",
       "8961        0  ...       0       0       0       0       0       0      160   \n",
       "2542        1  ...       0       0       0       0       0       0      189   \n",
       "8877        0  ...       0       0       1       0       1       0      156   \n",
       "\n",
       "       cytosine  guanine  thymine  \n",
       "14085        89      104      143  \n",
       "13466        99      103      141  \n",
       "8961         88       88      164  \n",
       "2542        102       90      119  \n",
       "8877         85      108      151  \n",
       "\n",
       "[5 rows x 4100 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Hex_train.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d80a654f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aaa</th>\n",
       "      <th>aac</th>\n",
       "      <th>aag</th>\n",
       "      <th>aat</th>\n",
       "      <th>aca</th>\n",
       "      <th>acc</th>\n",
       "      <th>acg</th>\n",
       "      <th>act</th>\n",
       "      <th>aga</th>\n",
       "      <th>agc</th>\n",
       "      <th>...</th>\n",
       "      <th>tgg</th>\n",
       "      <th>tgt</th>\n",
       "      <th>tta</th>\n",
       "      <th>ttc</th>\n",
       "      <th>ttg</th>\n",
       "      <th>ttt</th>\n",
       "      <th>adenine</th>\n",
       "      <th>cytosine</th>\n",
       "      <th>guanine</th>\n",
       "      <th>thymine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14085</th>\n",
       "      <td>22</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>15</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "      <td>18</td>\n",
       "      <td>164</td>\n",
       "      <td>89</td>\n",
       "      <td>104</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13466</th>\n",
       "      <td>19</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>18</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>157</td>\n",
       "      <td>99</td>\n",
       "      <td>103</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8961</th>\n",
       "      <td>28</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>17</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>16</td>\n",
       "      <td>18</td>\n",
       "      <td>160</td>\n",
       "      <td>88</td>\n",
       "      <td>88</td>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2542</th>\n",
       "      <td>34</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "      <td>19</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>189</td>\n",
       "      <td>102</td>\n",
       "      <td>90</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8877</th>\n",
       "      <td>17</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>14</td>\n",
       "      <td>16</td>\n",
       "      <td>156</td>\n",
       "      <td>85</td>\n",
       "      <td>108</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       aaa  aac  aag  aat  aca  acc  acg  act  aga  agc  ...  tgg  tgt  tta  \\\n",
       "14085   22    9   11   15    8    2    4   10   12    6  ...    8    5    7   \n",
       "13466   19    8   11   18    5    7    5    8    9    5  ...    6    8   14   \n",
       "8961    28    5    9   17    8    5    5    6   10    6  ...    6   10   12   \n",
       "2542    34   12   14   19   14    5    3    6   12    7  ...    7    1    6   \n",
       "8877    17    7   14   13    7    6    1    7   13    6  ...   10    7   10   \n",
       "\n",
       "       ttc  ttg  ttt  adenine  cytosine  guanine  thymine  \n",
       "14085    7   12   18      164        89      104      143  \n",
       "13466    5    4   11      157        99      103      141  \n",
       "8961    13   16   18      160        88       88      164  \n",
       "2542    10    7    9      189       102       90      119  \n",
       "8877     9   14   16      156        85      108      151  \n",
       "\n",
       "[5 rows x 68 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Cod_train.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a95562ff",
   "metadata": {},
   "source": [
    "### Scaling Hexamer and Codon Dataframes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ffc8866f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using MinMaxScaler to transform train, validate, test splits for hexamers\n",
    "scaler = MinMaxScaler()\n",
    "Hex_train_scaled = scaler.fit_transform(Hex_train)\n",
    "Hex_validate_scaled = scaler.fit_transform(Hex_validate)\n",
    "Hex_test_scaled = scaler.fit_transform(Hex_test)\n",
    "\n",
    "# Setting scaled hexamer df indices of dataframes to the train, validate, test indicies\n",
    "Hex_train_scaled = pd.DataFrame(Hex_train_scaled, columns = Hex_train.columns, index = Hex_train.index)\n",
    "Hex_validate_scaled = pd.DataFrame(Hex_validate_scaled, columns = Hex_validate.columns, index = Hex_validate.index)\n",
    "Hex_test_scaled = pd.DataFrame(Hex_test_scaled, columns = Hex_test.columns, index = Hex_test.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "68b21948",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aaaaaa</th>\n",
       "      <th>aaaaac</th>\n",
       "      <th>aaaaag</th>\n",
       "      <th>aaaaat</th>\n",
       "      <th>aaaaca</th>\n",
       "      <th>aaaacc</th>\n",
       "      <th>aaaacg</th>\n",
       "      <th>aaaact</th>\n",
       "      <th>aaaaga</th>\n",
       "      <th>aaaagc</th>\n",
       "      <th>...</th>\n",
       "      <th>ttttgg</th>\n",
       "      <th>ttttgt</th>\n",
       "      <th>ttttta</th>\n",
       "      <th>tttttc</th>\n",
       "      <th>tttttg</th>\n",
       "      <th>tttttt</th>\n",
       "      <th>adenine</th>\n",
       "      <th>cytosine</th>\n",
       "      <th>guanine</th>\n",
       "      <th>thymine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14085</th>\n",
       "      <td>0.023810</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.075</td>\n",
       "      <td>0.572222</td>\n",
       "      <td>0.322275</td>\n",
       "      <td>0.440476</td>\n",
       "      <td>0.414508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13466</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.25</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>0.369668</td>\n",
       "      <td>0.434524</td>\n",
       "      <td>0.404145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8961</th>\n",
       "      <td>0.119048</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.550000</td>\n",
       "      <td>0.317536</td>\n",
       "      <td>0.345238</td>\n",
       "      <td>0.523316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2542</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.25</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.711111</td>\n",
       "      <td>0.383886</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.290155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8877</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.527778</td>\n",
       "      <td>0.303318</td>\n",
       "      <td>0.464286</td>\n",
       "      <td>0.455959</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  4100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         aaaaaa    aaaaac  aaaaag    aaaaat    aaaaca  aaaacc  aaaacg  aaaact  \\\n",
       "14085  0.023810  0.166667   0.000  0.142857  0.000000    0.00     0.0    0.50   \n",
       "13466  0.000000  0.000000   0.000  0.000000  0.000000    0.00     0.0    0.00   \n",
       "8961   0.119048  0.166667   0.000  0.428571  0.000000    0.00     0.0    0.25   \n",
       "2542   0.000000  0.333333   0.000  0.428571  0.428571    0.25     0.0    0.25   \n",
       "8877   0.000000  0.000000   0.125  0.000000  0.142857    0.00     0.0    0.00   \n",
       "\n",
       "       aaaaga  aaaagc  ...  ttttgg  ttttgt    ttttta  tttttc    tttttg  \\\n",
       "14085   0.000    0.00  ...    0.25     0.0  0.000000     0.0  0.111111   \n",
       "13466   0.000    0.25  ...    0.00     0.0  0.000000     0.0  0.000000   \n",
       "8961    0.000    0.00  ...    0.00     0.0  0.000000     0.0  0.000000   \n",
       "2542    0.125    0.25  ...    0.00     0.0  0.000000     0.0  0.000000   \n",
       "8877    0.125    0.00  ...    0.00     0.0  0.166667     0.0  0.111111   \n",
       "\n",
       "       tttttt   adenine  cytosine   guanine   thymine  \n",
       "14085   0.075  0.572222  0.322275  0.440476  0.414508  \n",
       "13466   0.000  0.533333  0.369668  0.434524  0.404145  \n",
       "8961    0.000  0.550000  0.317536  0.345238  0.523316  \n",
       "2542    0.000  0.711111  0.383886  0.357143  0.290155  \n",
       "8877    0.000  0.527778  0.303318  0.464286  0.455959  \n",
       "\n",
       "[5 rows x 4100 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Hex_train_scaled.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "62a7c48d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using MinMaxScaler to transform train, validate, test splits for hexamers\n",
    "scaler = MinMaxScaler()\n",
    "Cod_train_scaled = scaler.fit_transform(Cod_train)\n",
    "Cod_validate_scaled = scaler.fit_transform(Cod_validate)\n",
    "Cod_test_scaled = scaler.fit_transform(Cod_test)\n",
    "\n",
    "# Setting scaled hexamer df indices of dataframes to the train, validate, test indicies\n",
    "Cod_train_scaled = pd.DataFrame(Cod_train_scaled, columns = Cod_train.columns, index = Cod_train.index)\n",
    "Cod_validate_scaled = pd.DataFrame(Cod_validate_scaled, columns = Cod_validate.columns, index = Cod_validate.index)\n",
    "Cod_test_scaled = pd.DataFrame(Cod_test_scaled, columns = Cod_test.columns, index = Cod_test.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "6b42a040",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aaa</th>\n",
       "      <th>aac</th>\n",
       "      <th>aag</th>\n",
       "      <th>aat</th>\n",
       "      <th>aca</th>\n",
       "      <th>acc</th>\n",
       "      <th>acg</th>\n",
       "      <th>act</th>\n",
       "      <th>aga</th>\n",
       "      <th>agc</th>\n",
       "      <th>...</th>\n",
       "      <th>tgg</th>\n",
       "      <th>tgt</th>\n",
       "      <th>tta</th>\n",
       "      <th>ttc</th>\n",
       "      <th>ttg</th>\n",
       "      <th>ttt</th>\n",
       "      <th>adenine</th>\n",
       "      <th>cytosine</th>\n",
       "      <th>guanine</th>\n",
       "      <th>thymine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14085</th>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.257143</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.173913</td>\n",
       "      <td>0.042553</td>\n",
       "      <td>0.210526</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>...</td>\n",
       "      <td>0.242424</td>\n",
       "      <td>0.098039</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.118644</td>\n",
       "      <td>0.196721</td>\n",
       "      <td>0.204545</td>\n",
       "      <td>0.572222</td>\n",
       "      <td>0.322275</td>\n",
       "      <td>0.440476</td>\n",
       "      <td>0.414508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13466</th>\n",
       "      <td>0.215909</td>\n",
       "      <td>0.228571</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.404762</td>\n",
       "      <td>0.108696</td>\n",
       "      <td>0.148936</td>\n",
       "      <td>0.263158</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.236842</td>\n",
       "      <td>0.217391</td>\n",
       "      <td>...</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>0.156863</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.084746</td>\n",
       "      <td>0.065574</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>0.369668</td>\n",
       "      <td>0.434524</td>\n",
       "      <td>0.404145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8961</th>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>0.173913</td>\n",
       "      <td>0.106383</td>\n",
       "      <td>0.263158</td>\n",
       "      <td>0.214286</td>\n",
       "      <td>0.263158</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>...</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>0.196078</td>\n",
       "      <td>0.244898</td>\n",
       "      <td>0.220339</td>\n",
       "      <td>0.262295</td>\n",
       "      <td>0.204545</td>\n",
       "      <td>0.550000</td>\n",
       "      <td>0.317536</td>\n",
       "      <td>0.345238</td>\n",
       "      <td>0.523316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2542</th>\n",
       "      <td>0.386364</td>\n",
       "      <td>0.342857</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.304348</td>\n",
       "      <td>0.106383</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>0.214286</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.304348</td>\n",
       "      <td>...</td>\n",
       "      <td>0.212121</td>\n",
       "      <td>0.019608</td>\n",
       "      <td>0.122449</td>\n",
       "      <td>0.169492</td>\n",
       "      <td>0.114754</td>\n",
       "      <td>0.102273</td>\n",
       "      <td>0.711111</td>\n",
       "      <td>0.383886</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.290155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8877</th>\n",
       "      <td>0.193182</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.152174</td>\n",
       "      <td>0.127660</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.342105</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>...</td>\n",
       "      <td>0.303030</td>\n",
       "      <td>0.137255</td>\n",
       "      <td>0.204082</td>\n",
       "      <td>0.152542</td>\n",
       "      <td>0.229508</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>0.527778</td>\n",
       "      <td>0.303318</td>\n",
       "      <td>0.464286</td>\n",
       "      <td>0.455959</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            aaa       aac   aag       aat       aca       acc       acg  \\\n",
       "14085  0.250000  0.257143  0.22  0.333333  0.173913  0.042553  0.210526   \n",
       "13466  0.215909  0.228571  0.22  0.404762  0.108696  0.148936  0.263158   \n",
       "8961   0.318182  0.142857  0.18  0.380952  0.173913  0.106383  0.263158   \n",
       "2542   0.386364  0.342857  0.28  0.428571  0.304348  0.106383  0.157895   \n",
       "8877   0.193182  0.200000  0.28  0.285714  0.152174  0.127660  0.052632   \n",
       "\n",
       "            act       aga       agc  ...       tgg       tgt       tta  \\\n",
       "14085  0.357143  0.315789  0.260870  ...  0.242424  0.098039  0.142857   \n",
       "13466  0.285714  0.236842  0.217391  ...  0.181818  0.156863  0.285714   \n",
       "8961   0.214286  0.263158  0.260870  ...  0.181818  0.196078  0.244898   \n",
       "2542   0.214286  0.315789  0.304348  ...  0.212121  0.019608  0.122449   \n",
       "8877   0.250000  0.342105  0.260870  ...  0.303030  0.137255  0.204082   \n",
       "\n",
       "            ttc       ttg       ttt   adenine  cytosine   guanine   thymine  \n",
       "14085  0.118644  0.196721  0.204545  0.572222  0.322275  0.440476  0.414508  \n",
       "13466  0.084746  0.065574  0.125000  0.533333  0.369668  0.434524  0.404145  \n",
       "8961   0.220339  0.262295  0.204545  0.550000  0.317536  0.345238  0.523316  \n",
       "2542   0.169492  0.114754  0.102273  0.711111  0.383886  0.357143  0.290155  \n",
       "8877   0.152542  0.229508  0.181818  0.527778  0.303318  0.464286  0.455959  \n",
       "\n",
       "[5 rows x 68 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Cod_train_scaled.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2fc09de",
   "metadata": {},
   "source": [
    "# Exploration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "4566b64a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>label</th>\n",
       "      <th>adenine</th>\n",
       "      <th>cytosine</th>\n",
       "      <th>guanine</th>\n",
       "      <th>thymine</th>\n",
       "      <th>hexamers</th>\n",
       "      <th>codons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14085</th>\n",
       "      <td>YPL026C_SKS1_501751</td>\n",
       "      <td>CATATAATATGAACTGCCAACGCACACATTTGGAGCTAGATATTTA...</td>\n",
       "      <td>1</td>\n",
       "      <td>164</td>\n",
       "      <td>89</td>\n",
       "      <td>104</td>\n",
       "      <td>143</td>\n",
       "      <td>[catata, atataa, tataat, ataata, taatat, aatat...</td>\n",
       "      <td>[cat, ata, tat, ata, taa, aat, ata, tat, atg, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13466</th>\n",
       "      <td>iYOR375C_1044573</td>\n",
       "      <td>AATCACAACCTATCTCTAGCCGCAAGATATATCTGTGTGAGAGGGG...</td>\n",
       "      <td>0</td>\n",
       "      <td>157</td>\n",
       "      <td>99</td>\n",
       "      <td>103</td>\n",
       "      <td>141</td>\n",
       "      <td>[aatcac, atcaca, tcacaa, cacaac, acaacc, caacc...</td>\n",
       "      <td>[aat, atc, tca, cac, aca, caa, aac, acc, cct, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        id                                           sequence  \\\n",
       "14085  YPL026C_SKS1_501751  CATATAATATGAACTGCCAACGCACACATTTGGAGCTAGATATTTA...   \n",
       "13466     iYOR375C_1044573  AATCACAACCTATCTCTAGCCGCAAGATATATCTGTGTGAGAGGGG...   \n",
       "\n",
       "       label  adenine  cytosine  guanine  thymine  \\\n",
       "14085      1      164        89      104      143   \n",
       "13466      0      157        99      103      141   \n",
       "\n",
       "                                                hexamers  \\\n",
       "14085  [catata, atataa, tataat, ataata, taatat, aatat...   \n",
       "13466  [aatcac, atcaca, tcacaa, cacaac, acaacc, caacc...   \n",
       "\n",
       "                                                  codons  \n",
       "14085  [cat, ata, tat, ata, taa, aat, ata, tat, atg, ...  \n",
       "13466  [aat, atc, tca, cac, aca, caa, aac, acc, cct, ...  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Taking a look at the explore data\n",
    "explore_train.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "1d06f949",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n</th>\n",
       "      <th>percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4576</td>\n",
       "      <td>0.559755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3599</td>\n",
       "      <td>0.440245</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      n   percent\n",
       "0  4576  0.559755\n",
       "1  3599  0.440245"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking percentage of labels in explore_train\n",
    "ex.show_counts_and_ratios(explore_train, 'label')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "7f91faa2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aaaaaa</th>\n",
       "      <th>aaaaac</th>\n",
       "      <th>aaaaag</th>\n",
       "      <th>aaaaat</th>\n",
       "      <th>aaaaca</th>\n",
       "      <th>aaaacc</th>\n",
       "      <th>aaaacg</th>\n",
       "      <th>aaaact</th>\n",
       "      <th>aaaaga</th>\n",
       "      <th>aaaagc</th>\n",
       "      <th>...</th>\n",
       "      <th>ttttgg</th>\n",
       "      <th>ttttgt</th>\n",
       "      <th>ttttta</th>\n",
       "      <th>tttttc</th>\n",
       "      <th>tttttg</th>\n",
       "      <th>tttttt</th>\n",
       "      <th>adenine</th>\n",
       "      <th>cytosine</th>\n",
       "      <th>guanine</th>\n",
       "      <th>thymine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>168</td>\n",
       "      <td>94</td>\n",
       "      <td>76</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>166</td>\n",
       "      <td>89</td>\n",
       "      <td>85</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>160</td>\n",
       "      <td>92</td>\n",
       "      <td>68</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>185</td>\n",
       "      <td>70</td>\n",
       "      <td>75</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>168</td>\n",
       "      <td>61</td>\n",
       "      <td>85</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  4100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   aaaaaa  aaaaac  aaaaag  aaaaat  aaaaca  aaaacc  aaaacg  aaaact  aaaaga  \\\n",
       "0       2       0       2       2       0       0       0       0       0   \n",
       "1       1       0       1       2       0       0       0       0       1   \n",
       "2       1       0       0       1       0       0       0       0       0   \n",
       "3      15       0       1       2       1       0       0       0       1   \n",
       "4       0       0       2       1       0       0       0       1       1   \n",
       "\n",
       "   aaaagc  ...  ttttgg  ttttgt  ttttta  tttttc  tttttg  tttttt  adenine  \\\n",
       "0       0  ...       0       0       0       3       0       1      168   \n",
       "1       0  ...       1       1       0       0       1       2      166   \n",
       "2       0  ...       1       1       0       0       0       0      160   \n",
       "3       0  ...       2       0       2       0       0       5      185   \n",
       "4       0  ...       1       2       2       2       0       7      168   \n",
       "\n",
       "   cytosine  guanine  thymine  \n",
       "0        94       76      162  \n",
       "1        89       85      160  \n",
       "2        92       68      180  \n",
       "3        70       75      170  \n",
       "4        61       85      186  \n",
       "\n",
       "[5 rows x 4100 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_hex.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "ae375f69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Vector</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4095</th>\n",
       "      <td>tttttt</td>\n",
       "      <td>27175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aaaaaa</td>\n",
       "      <td>26237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aaaaat</td>\n",
       "      <td>12010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1023</th>\n",
       "      <td>attttt</td>\n",
       "      <td>11913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4093</th>\n",
       "      <td>tttttc</td>\n",
       "      <td>11071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2048</th>\n",
       "      <td>gaaaaa</td>\n",
       "      <td>11060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>aagaaa</td>\n",
       "      <td>10827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4063</th>\n",
       "      <td>tttctt</td>\n",
       "      <td>10629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>aaagaa</td>\n",
       "      <td>10035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>agaaaa</td>\n",
       "      <td>9914</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Vector  Count\n",
       "4095  tttttt  27175\n",
       "0     aaaaaa  26237\n",
       "3     aaaaat  12010\n",
       "1023  attttt  11913\n",
       "4093  tttttc  11071\n",
       "2048  gaaaaa  11060\n",
       "128   aagaaa  10827\n",
       "4063  tttctt  10629\n",
       "32    aaagaa  10035\n",
       "512   agaaaa   9914"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking the number of occurrances of each vector\n",
    "Hex_vector_count = ex.vector_counts(X_hex.drop(columns = ['adenine', 'cytosine', 'guanine', 'thymine']))\n",
    "Hex_vector_count.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "8d707307",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Vector</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>ttt</td>\n",
       "      <td>308472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aaa</td>\n",
       "      <td>306545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aat</td>\n",
       "      <td>212453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>att</td>\n",
       "      <td>212037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ata</td>\n",
       "      <td>185429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>tat</td>\n",
       "      <td>184597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>gaa</td>\n",
       "      <td>173249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>ttc</td>\n",
       "      <td>171966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>taa</td>\n",
       "      <td>167218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>tta</td>\n",
       "      <td>166686</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Vector   Count\n",
       "63    ttt  308472\n",
       "0     aaa  306545\n",
       "3     aat  212453\n",
       "15    att  212037\n",
       "12    ata  185429\n",
       "51    tat  184597\n",
       "32    gaa  173249\n",
       "61    ttc  171966\n",
       "48    taa  167218\n",
       "60    tta  166686"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking the number of occurrances of each vector\n",
    "Cod_vector_count = ex.vector_counts(X_cod.drop(columns = ['adenine', 'cytosine', 'guanine', 'thymine']))\n",
    "Cod_vector_count.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39eb1e2c",
   "metadata": {},
   "source": [
    "# Modeling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "9f2c3b96",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55.98%\n"
     ]
    }
   ],
   "source": [
    "# Baseline based on most common label\n",
    "baseline = 0\n",
    "for label in y_train:\n",
    "    if label == 0:\n",
    "        baseline += 1\n",
    "\n",
    "base_acc = baseline / len(y_train)\n",
    "print(f'{round(base_acc * 100, 2)}%')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7defe35c",
   "metadata": {},
   "source": [
    "### Model baselines:\n",
    "> #### Based on Most Common Label:\n",
    "> - ~55.6%\n",
    "\n",
    "> #### Based on best preforming model http://www.jaist.ac.jp/~tran/nucleosome:\n",
    "![](JAIST_predictions.png)\n",
    "> - Best model for H4 dataset used K-Mer lengths of 4,5\n",
    "> - 88.09% \n",
    "> - Correlation coefficient = (0.76)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c71c2b9",
   "metadata": {},
   "source": [
    "## Hexamer Basic Classifier Models:\n",
    "- Exclusively Hexamer dataset\n",
    "- Basic models\n",
    "- Artificial Neural Network (ANN) models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "d61bcab5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Names of models to be used in classifier_models\n",
    "names = ['K Nearest Neighbors', 'Decision Tree', 'Random Forest', 'AddaBoost', \n",
    "         'Gaussian N-Bayes', 'Multinomial N-Bayes']\n",
    "\n",
    "# List of the classifiers to be used in modeling\n",
    "classifiers = [\n",
    "    KNeighborsClassifier(n_neighbors = 3),\n",
    "    DecisionTreeClassifier(max_depth = 5),\n",
    "    RandomForestClassifier(max_depth = 5, n_estimators = 10, max_features = 1 ),\n",
    "    AdaBoostClassifier(),\n",
    "    GaussianNB(),\n",
    "    MultinomialNB()\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "25635a24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "K Nearest Neighbors: Accuracy: 0.6200607206878286, Coeff: None\n",
      "Decision Tree: Accuracy: 0.7553556305045892, Coeff: None\n",
      "Random Forest: Accuracy: 0.5614678006781324, Coeff: None\n",
      "AddaBoost: Accuracy: 0.7975493262068574, Coeff: None\n",
      "Gaussian N-Bayes: Accuracy: 0.8321674502398603, Coeff: None\n",
      "Multinomial N-Bayes: Accuracy: 0.8459895616678587, Coeff: None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Multinomial N-Bayes</td>\n",
       "      <td>0.845990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gaussian N-Bayes</td>\n",
       "      <td>0.832167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AddaBoost</td>\n",
       "      <td>0.797549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Decision Tree</td>\n",
       "      <td>0.755356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K Nearest Neighbors</td>\n",
       "      <td>0.620061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Random Forest</td>\n",
       "      <td>0.561468</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Model  Accuracy\n",
       "5  Multinomial N-Bayes  0.845990\n",
       "4     Gaussian N-Bayes  0.832167\n",
       "3            AddaBoost  0.797549\n",
       "1        Decision Tree  0.755356\n",
       "0  K Nearest Neighbors  0.620061\n",
       "2        Random Forest  0.561468"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hex_train_metrics_df = md.classifier_models(Hex_train_scaled, y_train, names, classifiers)\n",
    "hex_train_metrics_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "2844df7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "K Nearest Neighbors: Accuracy: 0.5770492470492471, Coeff: None\n",
      "Decision Tree: Accuracy: 0.7442922262922264, Coeff: None\n",
      "Random Forest: Accuracy: 0.5650590150590152, Coeff: None\n",
      "AddaBoost: Accuracy: 0.7765641025641026, Coeff: None\n",
      "Gaussian N-Bayes: Accuracy: 0.8193365893365894, Coeff: None\n",
      "Multinomial N-Bayes: Accuracy: 0.8418876678876679, Coeff: None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Multinomial N-Bayes</td>\n",
       "      <td>0.841888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gaussian N-Bayes</td>\n",
       "      <td>0.819337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AddaBoost</td>\n",
       "      <td>0.776564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Decision Tree</td>\n",
       "      <td>0.744292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K Nearest Neighbors</td>\n",
       "      <td>0.577049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Random Forest</td>\n",
       "      <td>0.565059</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Model  Accuracy\n",
       "5  Multinomial N-Bayes  0.841888\n",
       "4     Gaussian N-Bayes  0.819337\n",
       "3            AddaBoost  0.776564\n",
       "1        Decision Tree  0.744292\n",
       "0  K Nearest Neighbors  0.577049\n",
       "2        Random Forest  0.565059"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Evaluate on out-of-sample validate data\n",
    "hex_val_metrics_df = md.classifier_models(Hex_validate_scaled, y_validate, names, classifiers)\n",
    "hex_val_metrics_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9f005db",
   "metadata": {},
   "source": [
    "## Codon Basic Classifier Models:\n",
    "- Exclusively Codon dataset\n",
    "- Basic models\n",
    "- Artificial Neural Network (ANN) models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "4c4d15a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "K Nearest Neighbors: Accuracy: 0.7905800636235497, Coeff: None\n",
      "Decision Tree: Accuracy: 0.747519998324121, Coeff: None\n",
      "Random Forest: Accuracy: 0.7190180845301404, Coeff: None\n",
      "AddaBoost: Accuracy: 0.8073396019188814, Coeff: None\n",
      "Gaussian N-Bayes: Accuracy: 0.7926572558079682, Coeff: None\n",
      "Multinomial N-Bayes: Accuracy: 0.6785309424126074, Coeff: None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AddaBoost</td>\n",
       "      <td>0.807340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gaussian N-Bayes</td>\n",
       "      <td>0.792657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K Nearest Neighbors</td>\n",
       "      <td>0.790580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Decision Tree</td>\n",
       "      <td>0.747520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Random Forest</td>\n",
       "      <td>0.719018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Multinomial N-Bayes</td>\n",
       "      <td>0.678531</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Model  Accuracy\n",
       "3            AddaBoost  0.807340\n",
       "4     Gaussian N-Bayes  0.792657\n",
       "0  K Nearest Neighbors  0.790580\n",
       "1        Decision Tree  0.747520\n",
       "2        Random Forest  0.719018\n",
       "5  Multinomial N-Bayes  0.678531"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Evaluate on train\n",
    "cod_train_metrics_df = md.classifier_models(Cod_train_scaled, y_train, names, classifiers)\n",
    "cod_train_metrics_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a12989fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "K Nearest Neighbors: Accuracy: 0.7668433048433049, Coeff: None\n",
      "Decision Tree: Accuracy: 0.7237370777370777, Coeff: None\n",
      "Random Forest: Accuracy: 0.7034831094831094, Coeff: None\n",
      "AddaBoost: Accuracy: 0.8065095645095646, Coeff: None\n",
      "Gaussian N-Bayes: Accuracy: 0.801949531949532, Coeff: None\n",
      "Multinomial N-Bayes: Accuracy: 0.6795002035002036, Coeff: None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AddaBoost</td>\n",
       "      <td>0.806510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gaussian N-Bayes</td>\n",
       "      <td>0.801950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K Nearest Neighbors</td>\n",
       "      <td>0.766843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Decision Tree</td>\n",
       "      <td>0.723737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Random Forest</td>\n",
       "      <td>0.703483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Multinomial N-Bayes</td>\n",
       "      <td>0.679500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Model  Accuracy\n",
       "3            AddaBoost  0.806510\n",
       "4     Gaussian N-Bayes  0.801950\n",
       "0  K Nearest Neighbors  0.766843\n",
       "1        Decision Tree  0.723737\n",
       "2        Random Forest  0.703483\n",
       "5  Multinomial N-Bayes  0.679500"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Evaluate on out-of-sample validate data\n",
    "cod_val_metrics_df = md.classifier_models(Cod_validate_scaled, y_validate, names, classifiers)\n",
    "cod_val_metrics_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7b4cbd9",
   "metadata": {},
   "source": [
    "## Hexamer Vector ANN\n",
    "#### Hex ANN Learning Classification Model 1: Basic Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "bd502d49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Converting df into list of arrays containing CountVectorized scaled values\n",
    "X_train = Hex_train_scaled.values\n",
    "X_validate = Hex_validate.values\n",
    "X_test = Hex_test_scaled.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "f4dd36cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Converting y_train/test to categorical matrices\n",
    "y_train_ = keras.utils.to_categorical(y_train, num_classes = 2)\n",
    "y_validate_ = keras.utils.to_categorical(y_validate, num_classes = 2)\n",
    "y_test_ = keras.utils.to_categorical(y_test, num_classes = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "9c7d9eda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(True, 4100, 8175)"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sanity checking\n",
    "len(X_train[0]) == len(Hex_train_scaled.columns), len(X_train[0]), len(X_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "466adfbb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([1., 0.], dtype=float32), 0)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sanity checking\n",
    "y_train_[1], explore_train.label.values[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "3494e68e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Make the basic ANN model with no hidden layers\n",
    "model0 = keras.Sequential([\n",
    "    keras.layers.Dense(2, input_dim = 4100, kernel_regularizer=l2(0.01), bias_regularizer=l2(0.01),activation = 'sigmoid')\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "061b04ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compile the model using categrical_crossentropy\n",
    "model0.compile(\n",
    "    optimizer = 'adam',\n",
    "    loss = 'categorical_crossentropy',\n",
    "    metrics = ['accuracy']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "459dca05",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/100\n",
      "256/256 [==============================] - 1s 1ms/step - loss: 0.5464 - accuracy: 0.8048\n",
      "Epoch 2/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4587 - accuracy: 0.8758\n",
      "Epoch 3/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4502 - accuracy: 0.8807\n",
      "Epoch 4/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4486 - accuracy: 0.8828\n",
      "Epoch 5/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4482 - accuracy: 0.8853\n",
      "Epoch 6/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4475 - accuracy: 0.8870\n",
      "Epoch 7/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4482 - accuracy: 0.8840\n",
      "Epoch 8/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4482 - accuracy: 0.8860: 0s - loss: 0.4458 - accuracy: \n",
      "Epoch 9/100\n",
      "256/256 [==============================] - 1s 2ms/step - loss: 0.4482 - accuracy: 0.8860\n",
      "Epoch 10/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4487 - accuracy: 0.8855\n",
      "Epoch 11/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4483 - accuracy: 0.8867\n",
      "Epoch 12/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4487 - accuracy: 0.8865\n",
      "Epoch 13/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4486 - accuracy: 0.8861\n",
      "Epoch 14/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4490 - accuracy: 0.8842\n",
      "Epoch 15/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4481 - accuracy: 0.8851\n",
      "Epoch 16/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4482 - accuracy: 0.8855\n",
      "Epoch 17/100\n",
      "256/256 [==============================] - 1s 2ms/step - loss: 0.4488 - accuracy: 0.8843\n",
      "Epoch 18/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4489 - accuracy: 0.8855\n",
      "Epoch 19/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4479 - accuracy: 0.8877\n",
      "Epoch 20/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4487 - accuracy: 0.8859\n",
      "Epoch 21/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4489 - accuracy: 0.8844\n",
      "Epoch 22/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4481 - accuracy: 0.8837\n",
      "Epoch 23/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4485 - accuracy: 0.8837\n",
      "Epoch 24/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4486 - accuracy: 0.8828\n",
      "Epoch 25/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4486 - accuracy: 0.8851\n",
      "Epoch 26/100\n",
      "256/256 [==============================] - 1s 2ms/step - loss: 0.4488 - accuracy: 0.8853\n",
      "Epoch 27/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4487 - accuracy: 0.8845\n",
      "Epoch 28/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4480 - accuracy: 0.8880\n",
      "Epoch 29/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4489 - accuracy: 0.8844\n",
      "Epoch 30/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4489 - accuracy: 0.8850\n",
      "Epoch 31/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4484 - accuracy: 0.8842\n",
      "Epoch 32/100\n",
      "256/256 [==============================] - 1s 2ms/step - loss: 0.4483 - accuracy: 0.8855\n",
      "Epoch 33/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4495 - accuracy: 0.8848\n",
      "Epoch 34/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4484 - accuracy: 0.8828\n",
      "Epoch 35/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4483 - accuracy: 0.8865\n",
      "Epoch 36/100\n",
      "256/256 [==============================] - 1s 2ms/step - loss: 0.4483 - accuracy: 0.8834\n",
      "Epoch 37/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4483 - accuracy: 0.8864\n",
      "Epoch 38/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4483 - accuracy: 0.8838\n",
      "Epoch 39/100\n",
      "256/256 [==============================] - 1s 2ms/step - loss: 0.4487 - accuracy: 0.8856\n",
      "Epoch 40/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4488 - accuracy: 0.8834\n",
      "Epoch 41/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4489 - accuracy: 0.8849\n",
      "Epoch 42/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4486 - accuracy: 0.8869\n",
      "Epoch 43/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4492 - accuracy: 0.8823\n",
      "Epoch 44/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4489 - accuracy: 0.8846\n",
      "Epoch 45/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4484 - accuracy: 0.8821\n",
      "Epoch 46/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4482 - accuracy: 0.8866\n",
      "Epoch 47/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4496 - accuracy: 0.8829\n",
      "Epoch 48/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4487 - accuracy: 0.8837\n",
      "Epoch 49/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4484 - accuracy: 0.8866\n",
      "Epoch 50/100\n",
      "256/256 [==============================] - 1s 2ms/step - loss: 0.4488 - accuracy: 0.8846\n",
      "Epoch 51/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4488 - accuracy: 0.8849\n",
      "Epoch 52/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4489 - accuracy: 0.8848\n",
      "Epoch 53/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4484 - accuracy: 0.8838\n",
      "Epoch 54/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4493 - accuracy: 0.8853\n",
      "Epoch 55/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4486 - accuracy: 0.8848\n",
      "Epoch 56/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4484 - accuracy: 0.8851\n",
      "Epoch 57/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4489 - accuracy: 0.8855\n",
      "Epoch 58/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4485 - accuracy: 0.8869\n",
      "Epoch 59/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4489 - accuracy: 0.8850\n",
      "Epoch 60/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4483 - accuracy: 0.8861\n",
      "Epoch 61/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4495 - accuracy: 0.8844\n",
      "Epoch 62/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4487 - accuracy: 0.8860\n",
      "Epoch 63/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4486 - accuracy: 0.8860\n",
      "Epoch 64/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4486 - accuracy: 0.8860\n",
      "Epoch 65/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4490 - accuracy: 0.8838\n",
      "Epoch 66/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4487 - accuracy: 0.8848\n",
      "Epoch 67/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4486 - accuracy: 0.8835\n",
      "Epoch 68/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4483 - accuracy: 0.8851\n",
      "Epoch 69/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4482 - accuracy: 0.8881\n",
      "Epoch 70/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4485 - accuracy: 0.8861\n",
      "Epoch 71/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4485 - accuracy: 0.8837\n",
      "Epoch 72/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4489 - accuracy: 0.8849\n",
      "Epoch 73/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4482 - accuracy: 0.8846\n",
      "Epoch 74/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4486 - accuracy: 0.8853\n",
      "Epoch 75/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4489 - accuracy: 0.8846\n",
      "Epoch 76/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4495 - accuracy: 0.8835\n",
      "Epoch 77/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4485 - accuracy: 0.8851\n",
      "Epoch 78/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4480 - accuracy: 0.8838: 0s - loss: 0.4438 - accuracy: \n",
      "Epoch 79/100\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4485 - accuracy: 0.8843\n",
      "Epoch 80/100\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "256/256 [==============================] - 0s 2ms/step - loss: 0.4495 - accuracy: 0.8839\n",
      "Epoch 81/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4487 - accuracy: 0.8856\n",
      "Epoch 82/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4492 - accuracy: 0.8827\n",
      "Epoch 83/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4484 - accuracy: 0.8854\n",
      "Epoch 84/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4481 - accuracy: 0.8853\n",
      "Epoch 85/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4491 - accuracy: 0.8826\n",
      "Epoch 86/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4485 - accuracy: 0.8853\n",
      "Epoch 87/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4494 - accuracy: 0.8860\n",
      "Epoch 88/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4477 - accuracy: 0.8876\n",
      "Epoch 89/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4487 - accuracy: 0.8862\n",
      "Epoch 90/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4488 - accuracy: 0.8848\n",
      "Epoch 91/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4492 - accuracy: 0.8837\n",
      "Epoch 92/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4488 - accuracy: 0.8846\n",
      "Epoch 93/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4492 - accuracy: 0.8823\n",
      "Epoch 94/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4486 - accuracy: 0.8865\n",
      "Epoch 95/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4483 - accuracy: 0.8855\n",
      "Epoch 96/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4483 - accuracy: 0.8871\n",
      "Epoch 97/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4482 - accuracy: 0.8865\n",
      "Epoch 98/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4492 - accuracy: 0.8857\n",
      "Epoch 99/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4476 - accuracy: 0.8860\n",
      "Epoch 100/100\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.4487 - accuracy: 0.8870\n"
     ]
    }
   ],
   "source": [
    "# Fitting the model\n",
    "train_scores = model0.fit(X_train, y_train_, epochs = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "9fe941d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Appending train metrics df with model score\n",
    "hex_train_metrics_df = hex_train_metrics_df.append({'Model': 'ANN 1', 'Accuracy': train_scores.history['accuracy'][-1]}, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "bea52067",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "110/110 [==============================] - 0s 1ms/step - loss: 15.3313 - accuracy: 0.5602\n"
     ]
    }
   ],
   "source": [
    "val_scores = model0.evaluate(X_validate, y_validate_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "92e365fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Appending validate metrics df with model score\n",
    "hex_val_metrics_df = hex_val_metrics_df.append({'Model': 'ANN 1', 'Accuracy': val_scores[1]}, ignore_index = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0887ff35",
   "metadata": {},
   "source": [
    "#### Hex ANN Learning Classification Model 2: 1 Hidden Layer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "8d0ee85e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the ANN model with a hidden layer\n",
    "model1 = keras.Sequential([\n",
    "    keras.layers.Dense(2919, input_dim = 4100, activation = 'relu'),\n",
    "    keras.layers.Dense(100, activation = 'relu'),\n",
    "    keras.layers.Dense(2, activation = 'sigmoid')\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "f2d97fc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compile the model using categrical_crossentropy\n",
    "model1.compile(\n",
    "    optimizer = 'adam',\n",
    "    loss = 'categorical_crossentropy',\n",
    "    metrics = ['accuracy']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "121ae90d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/5\n",
      "256/256 [==============================] - 11s 43ms/step - loss: 0.3979 - accuracy: 0.8296\n",
      "Epoch 2/5\n",
      "256/256 [==============================] - 11s 42ms/step - loss: 0.1986 - accuracy: 0.9228\n",
      "Epoch 3/5\n",
      "256/256 [==============================] - 12s 46ms/step - loss: 0.0584 - accuracy: 0.9806\n",
      "Epoch 4/5\n",
      "256/256 [==============================] - 12s 47ms/step - loss: 0.0228 - accuracy: 0.9929\n",
      "Epoch 5/5\n",
      "256/256 [==============================] - 11s 43ms/step - loss: 0.0104 - accuracy: 0.9969\n"
     ]
    }
   ],
   "source": [
    "# Fitting the model\n",
    "train_scores = model1.fit(X_train, y_train_, epochs = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "50df15d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Appending train metrics df with model score\n",
    "hex_train_metrics_df = hex_train_metrics_df.append({'Model': 'ANN 2', 'Accuracy': train_scores.history['accuracy'][-1]}, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "1f860ed2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "110/110 [==============================] - 2s 15ms/step - loss: 3.1464 - accuracy: 0.8191\n"
     ]
    }
   ],
   "source": [
    "val_scores = model1.evaluate(X_validate, y_validate_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "a694d17b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Appending validate metrics df with model score\n",
    "hex_val_metrics_df = hex_val_metrics_df.append({'Model': 'ANN 2', 'Accuracy': val_scores[1]}, ignore_index = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1653dfcc",
   "metadata": {},
   "source": [
    "### Codon Vector ANN:\n",
    "#### Codon ANN Learning Classification Model 1: Basic Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "75ebf47a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Converting df into list of arrays containing CountVectorized scaled values\n",
    "X_train = Cod_train_scaled.values\n",
    "X_validate = Cod_validate.values\n",
    "X_test = Cod_test_scaled.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "06772cdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(True, 68, 8175)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sanity checking\n",
    "len(X_train[0]) == len(Cod_train_scaled.columns), len(X_train[0]), len(X_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "43c3d2d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the basic ANN model with no hidden layers\n",
    "model0 = keras.Sequential([\n",
    "    keras.layers.Dense(2, input_dim = 68, kernel_regularizer=l2(0.01), bias_regularizer=l2(0.01),activation = 'sigmoid')\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "e654d71a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compile the model using categrical_crossentropy\n",
    "model0.compile(\n",
    "    optimizer = 'adam',\n",
    "    loss = 'categorical_crossentropy',\n",
    "    metrics = ['accuracy']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "bac46050",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/25\n",
      "256/256 [==============================] - 0s 777us/step - loss: 0.6861 - accuracy: 0.6275\n",
      "Epoch 2/25\n",
      "256/256 [==============================] - 0s 749us/step - loss: 0.6328 - accuracy: 0.7187\n",
      "Epoch 3/25\n",
      "256/256 [==============================] - 0s 754us/step - loss: 0.6061 - accuracy: 0.7640\n",
      "Epoch 4/25\n",
      "256/256 [==============================] - 0s 773us/step - loss: 0.5919 - accuracy: 0.7829\n",
      "Epoch 5/25\n",
      "256/256 [==============================] - 0s 752us/step - loss: 0.5848 - accuracy: 0.7938\n",
      "Epoch 6/25\n",
      "256/256 [==============================] - 0s 776us/step - loss: 0.5809 - accuracy: 0.7969\n",
      "Epoch 7/25\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.5787 - accuracy: 0.8045\n",
      "Epoch 8/25\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.5779 - accuracy: 0.8050\n",
      "Epoch 9/25\n",
      "256/256 [==============================] - 0s 801us/step - loss: 0.5771 - accuracy: 0.8069\n",
      "Epoch 10/25\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.5769 - accuracy: 0.8088\n",
      "Epoch 11/25\n",
      "256/256 [==============================] - 0s 873us/step - loss: 0.5766 - accuracy: 0.8062\n",
      "Epoch 12/25\n",
      "256/256 [==============================] - 0s 863us/step - loss: 0.5766 - accuracy: 0.8102\n",
      "Epoch 13/25\n",
      "256/256 [==============================] - 0s 901us/step - loss: 0.5764 - accuracy: 0.8088\n",
      "Epoch 14/25\n",
      "256/256 [==============================] - 0s 855us/step - loss: 0.5761 - accuracy: 0.8100\n",
      "Epoch 15/25\n",
      "256/256 [==============================] - 0s 852us/step - loss: 0.5759 - accuracy: 0.8098\n",
      "Epoch 16/25\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.5764 - accuracy: 0.8102\n",
      "Epoch 17/25\n",
      "256/256 [==============================] - 0s 907us/step - loss: 0.5761 - accuracy: 0.8104\n",
      "Epoch 18/25\n",
      "256/256 [==============================] - 0s 861us/step - loss: 0.5762 - accuracy: 0.8121\n",
      "Epoch 19/25\n",
      "256/256 [==============================] - 0s 837us/step - loss: 0.5758 - accuracy: 0.8103\n",
      "Epoch 20/25\n",
      "256/256 [==============================] - 0s 838us/step - loss: 0.5758 - accuracy: 0.8067\n",
      "Epoch 21/25\n",
      "256/256 [==============================] - 0s 854us/step - loss: 0.5764 - accuracy: 0.8097\n",
      "Epoch 22/25\n",
      "256/256 [==============================] - 0s 901us/step - loss: 0.5758 - accuracy: 0.8116\n",
      "Epoch 23/25\n",
      "256/256 [==============================] - 0s 927us/step - loss: 0.5763 - accuracy: 0.8086\n",
      "Epoch 24/25\n",
      "256/256 [==============================] - 0s 809us/step - loss: 0.5763 - accuracy: 0.8103\n",
      "Epoch 25/25\n",
      "256/256 [==============================] - 0s 837us/step - loss: 0.5762 - accuracy: 0.8084\n"
     ]
    }
   ],
   "source": [
    "# Fitting the model\n",
    "train_scores = model0.fit(X_train, y_train_, epochs = 25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "73679c37",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Appending train metrics df with model score\n",
    "cod_train_metrics_df = cod_train_metrics_df.append({'Model': 'ANN 1', 'Accuracy': train_scores.history['accuracy'][-1]}, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "12f818c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "110/110 [==============================] - 0s 1ms/step - loss: 40.0140 - accuracy: 0.5608\n"
     ]
    }
   ],
   "source": [
    "val_scores = model0.evaluate(X_validate, y_validate_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "ef1f97d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Appending validate metrics df with model score\n",
    "cod_val_metrics_df = cod_val_metrics_df.append({'Model': 'ANN 1', 'Accuracy': val_scores[1]}, ignore_index = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f7ffb7a",
   "metadata": {},
   "source": [
    "#### Codon ANN Learning Classification Model 2: 1 Hidden Layer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "31169da6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the ANN model with a hidden layer\n",
    "model1 = keras.Sequential([\n",
    "    keras.layers.Dense(60, input_dim = 68, activation = 'relu'),\n",
    "    keras.layers.Dense(30, activation = 'relu'),\n",
    "    keras.layers.Dense(2, activation = 'sigmoid')\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "b9c0ddd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compile the model using categrical_crossentropy\n",
    "model1.compile(\n",
    "    optimizer = 'adam',\n",
    "    loss = 'categorical_crossentropy',\n",
    "    metrics = ['accuracy']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "8a81ed1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/5\n",
      "256/256 [==============================] - 1s 1ms/step - loss: 0.5124 - accuracy: 0.7644\n",
      "Epoch 2/5\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.3976 - accuracy: 0.8296\n",
      "Epoch 3/5\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.3847 - accuracy: 0.8415\n",
      "Epoch 4/5\n",
      "256/256 [==============================] - 0s 2ms/step - loss: 0.3747 - accuracy: 0.8415\n",
      "Epoch 5/5\n",
      "256/256 [==============================] - 0s 1ms/step - loss: 0.3695 - accuracy: 0.8431\n"
     ]
    }
   ],
   "source": [
    "# Fitting the model\n",
    "train_scores = model1.fit(X_train, y_train_, epochs = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "a031e56d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Appending train metrics df with model score\n",
    "cod_train_metrics_df = hex_train_metrics_df.append({'Model': 'ANN 2', 'Accuracy': train_scores.history['accuracy'][-1]}, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "a30c01ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "110/110 [==============================] - 0s 2ms/step - loss: 87.5675 - accuracy: 0.5634\n"
     ]
    }
   ],
   "source": [
    "val_scores = model1.evaluate(X_validate, y_validate_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "f9a471f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Appending validate metrics df with model score\n",
    "cod_val_metrics_df = cod_val_metrics_df.append({'Model': 'ANN 2', 'Accuracy': val_scores[1]}, ignore_index = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c494fb95",
   "metadata": {},
   "source": [
    "## Out-of-sample (Validate) Metrics:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "3bda6732",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Multinomial N-Bayes</td>\n",
       "      <td>0.841888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gaussian N-Bayes</td>\n",
       "      <td>0.819337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AddaBoost</td>\n",
       "      <td>0.776564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Decision Tree</td>\n",
       "      <td>0.744292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K Nearest Neighbors</td>\n",
       "      <td>0.577049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Random Forest</td>\n",
       "      <td>0.565059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ANN 1</td>\n",
       "      <td>0.560217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ANN 2</td>\n",
       "      <td>0.819064</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Model  Accuracy\n",
       "0  Multinomial N-Bayes  0.841888\n",
       "1     Gaussian N-Bayes  0.819337\n",
       "2            AddaBoost  0.776564\n",
       "3        Decision Tree  0.744292\n",
       "4  K Nearest Neighbors  0.577049\n",
       "5        Random Forest  0.565059\n",
       "6                ANN 1  0.560217\n",
       "7                ANN 2  0.819064"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hex_val_metrics_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "b24271c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AddaBoost</td>\n",
       "      <td>0.806510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gaussian N-Bayes</td>\n",
       "      <td>0.801950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K Nearest Neighbors</td>\n",
       "      <td>0.766843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Decision Tree</td>\n",
       "      <td>0.723737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Random Forest</td>\n",
       "      <td>0.703483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Multinomial N-Bayes</td>\n",
       "      <td>0.679500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ANN 1</td>\n",
       "      <td>0.560788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ANN 2</td>\n",
       "      <td>0.563356</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Model  Accuracy\n",
       "0            AddaBoost  0.806510\n",
       "1     Gaussian N-Bayes  0.801950\n",
       "2  K Nearest Neighbors  0.766843\n",
       "3        Decision Tree  0.723737\n",
       "4        Random Forest  0.703483\n",
       "5  Multinomial N-Bayes  0.679500\n",
       "6                ANN 1  0.560788\n",
       "7                ANN 2  0.563356"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cod_val_metrics_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6b70b52",
   "metadata": {},
   "source": [
    "### Pre-test takeaways:\n",
    "- Multinomial Naive Bayes on the Hexamer data preformed the best of my models\n",
    "    - Will use this model to evaluate on test\n",
    "- All of the ANN models were considerably overfit\n",
    "    - ANN 2 Preformed best of the ANN models on both Codon and Hexamer vector datasets\n",
    "\n",
    "## Out-of-sample (Test) Evaluation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "8c25074f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Multinomial N-Bayes: Accuracy: 0.8438356164383561, Coeff: None\n"
     ]
    }
   ],
   "source": [
    "# Evaluate Multinomial Naive Bayes on Test Data\n",
    "test_metric_df = md.classifier_models(Hex_test_scaled, y_test, ['Multinomial N-Bayes'], [MultinomialNB()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "e41f3821",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Multinomial N-Bayes</td>\n",
       "      <td>0.843836</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Model  Accuracy\n",
       "0  Multinomial N-Bayes  0.843836"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_metric_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efbc2a7c",
   "metadata": {},
   "source": [
    "# Conclusion:\n",
    "> ## Modeling:\n",
    "> - Best preforming model outpreformed baseline by almost 30 points on test data\n",
    "> - Best model didnt beat preformance of http://www.jaist.ac.jp/~tran/nucleosome:\n",
    ">     - My model lost by ~4 points\n",
    "\n",
    "> ## Key Takeaways:\n",
    "> - Overall, my ANN models seemed to rank in the middle/end of my models overall.\n",
    "> - I achieved a ~84.4% accuracy on test data using a Multinomial Naive Bayes test.\n",
    "\n",
    "> ## Reccomendations:\n",
    "> - I reccomend that the model used by www.jaist.ac be used in lieu of my own models. Additionally, the hexamer dataset seemed to be more indicative of positive or negative labels than the codon dataset. If pursuing the ANN models further I reccomend using a model with at least one hidden layer as it outpreformed the single layer model slightly with the codon dataset and significantly improved with the hexamer dataset.\n",
    "\n",
    "> ## Future Iterations:\n",
    "> For any future iterations I would like to further develop my ANN models to try and minimize the prevalent overfitting issue present in my ANN models. I would like to try different vector lengths instead of 3 and 6 and would try to maximize model preformance with further tinkering with model hyperparameters. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76707ff1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
